```{r setup}
#| include: false
schtools::set_knitr_opts()
```

# Introduction

## Microbial communities in human health

Microbial communities are assemblages of microorganisms -- archaea, bacteria,
fungi, protists, and viruses -- that inhabit a local environment [@konopka_what_2009].
A microbiome consists of the microbial community in its environment together
with the molecules they produce such as nucleic acids, proteins, lipids,
metabolites, and more [@berg_microbiome_2020].
Microbiomes are thus tightly associated with the local environment they occupy
such as waterways, soil layers, ocean floors, plants, insects, and animals.
The human body hosts microbes that inhabit the skin, mouth, gut, vagina, airway,
and other body regions [@human_microbiome_project_consortium_structure_2012].
The living members of a microbiome or referred to as the microbiota, and they
interact with each other cooperatively and competitively [@coyte_understanding_2019]
and can influence the health of the host directly or indirectly [@tremaroli_functional_2012].
The composition of a microbiome can be characterized according to the taxonomy
of its microbiota, the functions they carry out, and the metabolites they
produce.

High throughput sequencing and other 'omics techniques can be used to
characterize the metagenomes, metatranscriptomes, metaproteomes, and
meta-metabolomes of microbiomes and describe how they change over time,
in response to changing environments,
or between healthy and diseased states of the host.
A benefit of 'omics techniques is they do not require microbes to be cultured in
a laboratory, making it possible to observe genes, gene products, and
metabolites from microbes missed by culturing techniques and even discover new species
[@dickson_analysis_2020].
These large, multivariate datasets present a challenge for bioinformatic
analysis, as greater computational resources and more sophisticated statistical
techniques are required to process and analyze big data
[@human_microbiome_project_consortium_structure_2012].
An older but still widely-used method to profile the taxonomic composition of
microbial communities is amplicon sequencing.
In amplicon sequencing, a region of a marker gene is selected depending on which
domain of microbial life is being targeted, and the DNA sequences matching that
region are amplified and sequenced [@pace_analysis_1986;@kolbert_ribosomal_1999].
Relative to shotgun metagenomics, amplicon sequencing significantly reduces the
costs and computational resources needed to characterize the taxonomic
composition of microbiomes.
Amplicon sequence data can typically be identified at the genus level, while
shotgun metagenomics data provide a finer taxonomic resolution at the species
or strain level [@schloss_reintroducing_2019].
When researchers need to extract sequences from many samples and only require
taxonomic resolution at the genus level, amplicon sequencing is a practical choice.

### _Clostridioides difficile_ infection

The healthy gut microbiome is resistant to colonization and infection from
pathogens due to competition from beneficial microbes.
Medications such as antibiotics, proton-pump inhibitors, and osmotic laxatives
can disrupt the taxonomic and functional composition of the gut microbiome,
thereby allowing pathogens to gain a foothold
[@janarthanan_clostridium_2012;@tomkovich_osmotic_2021].
_C. difficile_ is classically considered a hospital-acquired pathogen that
infects patients who are taking antibiotics for other illnesses, and especially
in elderly patients as the immune system weakens with age.
However, community-acquired _C. difficile_ infection (CDI) is increasing, even
in patients with no recent history of antibiotic use [@feuerstadt_burden_2023].
One mechanism through which the healthy gut community protects against CDI is
bile acid metabolism, which can inhibit the growth and alter toxin production of
_C. difficile_
[@theriot_antibiotic-induced_2016;@buffie_precision_2015;@sorg_inhibiting_2010].
Mouse studies have shown that the initial taxonomic composition of the gut
microbiome can influence _C. difficile_ clearance, host moribundity, and cecal
tissue damage in infected mice
[@tomkovich_initial_2020;@lesniak_clearance_2021;@lesniak_gut_2022].
Differences between resilient versus susceptible microbiomes could be used as
biomarkers to identify patients at risk of being infected or developing severe
complications.

Clinical outcomes of CDI can be severe, as a small portion of patients
experience complications requiring ICU admission due to CDI such as ileus,
toxic megacolon, or death in 8-9% of cases [@lucado_clostridium_2012;@kassam_clostridium_2016].
Colectomy is used as a last-resort treatment to prevent death when other
treatments fail, and the mortality rate in patients who undergo colectomy for
CDI is approximately 35% [@peprah_comparison_2019].
CDI is of particular cause for concern due to the risk of recurrence, where a
patient experiences another CDI within 2-8 weeks of a prior CDI [@kelly_can_2012].
It is thought that the antibiotics prescribed to treat CDI may also prevent the
microbiome from recovering, thereby perpetuating a cycle of perturbance and
_C. difficile_ proliferation.
Adjuvant therapies such as bezlotoxumab and fecal microbiota transplant have
recently been introduced to help break the cycle of recurrence by targeting the
toxins produced by _C. difficile_ or restoring the microbiota to a healthy
state, but they are not yet used widely and typically only in patients
experiencing a first or second recurrence [@de_la_villa_real-world_2023].
Furthermore, vancomycin is commonly used to treat first cases of CDI, but
vancomycin-resistant _Enterococcus_ is becoming more common and enterococci have
been shown to exacerbate the pathogenesis of _C. difficile_
[@poduval_clostridium_2000;@smith_enterococci_2022].
There is a great need for improved therapies to prevent recurrent and severe
CDI, as well as for tools to predict which patients are at risk so that
clinicians can adjust treatment plans to prevent adverse outcomes from
occurring.

## Machine learning for science and health care

Machine learning techniques applied to large datasets have transformed the
quantitative sciences towards a data-driven paradigm.
Supervised ML approaches can be used to classify samples or make predictions,
and researchers often claim that the good discriminative performance of a model
supports the veracity of an underlying scientific claim [@kapoor_leakage_2022].
Alternatively, ML can be used pragmatically in fields like medicine in order to
aid in diagnosing diseases or to make predictions about disease outcomes,
with the goal of improving health care.
Models based on clinical laboratory tests and electronic health record (EHR)
data and have been trained to predict deterioration in COVID-19 patients
[@kamran_early_2022], identify patients at risk of being infected with _C.
difficile_ in ICU wards [@otles_clostridioides_2023], and predict outcomes in
CDI patients [@dieterle_systemic_2020].
ML models trained on microbiota have been used to improve detection of
colorectal cancer [@baxter_microbiota-based_2016], distinguish CDI patients from
diarrheal controls [@schubert_microbiome_2014], and identify which members of
the microbiota contribute most to the performance of these models
[@topcuoglu_framework_2020].

While ML techniques hold great promise to improve health care,
caution must be taken to train, test, validate, and deploy ML models responsibly.
Pervasive pitfalls have been identified in studies applying ML; these
include data leakage between the training and test/validation set,
failing to set a random seed,
biased training data,
inappropriate choice of performance metric,
not reporting variation in performance,
and neglecting to document methods in sufficient detail
[@kapoor_leakage_2022;@topcuoglu_framework_2020].
Errors in ML can invalidate the conclusions of a study and erode trust in the
scientific endeavor.
Even worse, errors can be dangerous when ML is applied to health care, as
diagnoses and prognoses assisted by ML can directly affect patients
[@beam_challenges_2020].
Before a model can ever be deployed in clinical practice, it must be rigorously
evaluated to ensure it avoids these errors, and ultimately that it will improve
rather than worsen patient outcomes [@wiens_no_2019].
ML practitioners must take care to avoid technical pitfalls and engage with
experts from interdisciplinary fields to ensure their models will be useful and
beneficial for clinical practice.

## Democratizing reproducible data science

Data science is an interdisciplinary field that integrates computer science,
statistics, and expertise from a problem domain.
When the problem domain is biology, the field could be referred to as
biological data science, computational biology, or bioinformatics, although
there is no consensus definition for any of these terms.
As costs decrease for generating large datasets such as those from
high-throughput sequencing experiments, there is an ever-growing need for data
science practitioners with the skills and knowledge to process the data,
make inferences, and communicate their findings.
Democratizing data science means making the theory, methods, and tools, more
accessible by creating educational resources, user-friendly software tools, or
even simply making data publicly available.
Accessibility is important for filling the growing demand for skilled data
scientists across sectors as well as to improve diversity in the field
[@ncses_diversity_2023;@gilshan_ethics_2021;@cowgill_biased_2020;@hong_groups_2004].
Non-profit organizations have been founded to help address the diversity gap in
computer science, data science, and other STEM fields including
Girls Who Code,
Women in Science and Engineering,
Association for Women in Science,
Society for Advancement of Chicanos and Native Americans in Science,
and many others.

An important attribute of any scientific finding is reproducibility, where
others can repeat the same methods on the original dataset to obtain the same
result [@schloss_identifying_2018].
Reproducibility does not guarantee correctness, replicability, nor
generalizability, but it is a minimal achievable standard that helps others
evaluate scientific claims [@sandve_ten_2013].
A finding could be entirely unreproducible, where the data are not
shared and the analysis methods are not described in sufficient detail.
Achieving perfect reproduciblity is unlikely as eventually link rot, software
bugs, and shuttering of organizations can occur, but "good enough" practices are
attainable [@wilson_good_2017].
Aside from enabling others to validate or build upon one's work, reproducible
practices make researchers more productive both collaboratively and individually
[@wilson_good_2017].
As a reproducibility crisis has been identified in virtually all scientific
fields including microbiology, bioinformatics, and ML for health research,
promoting and encouraging reproducible practices is important to re-establish
trust and trustworthiness in scientific findings
[@gelman_statistical_2016;@schloss_identifying_2018;@kapoor_leakage_2022;@mcdermott_reproducibility_2021].
However, many early-career researchers lack the quantitative and computational
skills and self-confidence necessary to perform reproducible computational
science, in some cases due to prior demotivating experiences
[@cuddington_challenges_2023].
Toward the goal of disseminating reproducible research practices,
Software Carpentry, Data Carpentry, and Library Carpentry (together under the
umbrella term "The Carpentries") have developed extensive educational
materials and taught them in hands-on workshops worldwide to researchers,
scientists, librarians, and other data wranglers [@wilson_software_2016].
A particularly important contribution of The Carpentries is the instructor
training course which promotes evidence-based pedagogical practices that
motivate and empower learners, such as instruction via participatory live-coding
[@deans_science_2015;@carpentries_training_2021].
Improving access to educational resources for best practices in coding and
data science will equip budding scientists with the skills necessary to
conduct and communicate their work reproducibly.

<!--
ml common pitfalls.
is democratizing data science dangerous?
-->

## Dissertation outline and contributions

In the preambles of Chapters 2 through 4, I note my specific contributions
to the work described in each chapter.
Chapter 2 presents OptiFit, a new OTU clustering algorithm that enables
researchers to fit new data to existing \textit{de novo} OTUs while maintaining
OTU quality.
Chapter 3 presents findings from training ML models to predict the severity of
CDI from OTUs and comparing model performance to prior approaches.
Chapter 4 introduces two curricula and one software package which help
democratize data science for a range of audiences.
In Chapter 5, I discuss the impacts of the findings presented in Chapters 2
through 4 and propose future work to build upon this dissertation.

## Datasets used in this dissertation

In Chapter 2, we re-used previously published 16S rRNA gene amplicon sequence
data extracted from four different communities: soil, marine, mouse gut, and
human gut.
Using multiple datasets from disparate sources allowed us to
demonstrate the suitability of OptiFit for microbiome researchers and microbial
ecologists with diverse scientific interests.
In Chapter 3, we used a dataset of 16S rRNA gene amplicon sequences extracted
from 1,277 stool samples collected on the day of diagnosis from
CDI patients at the University of Michigan.
White blood cell counts and creatinine levels were also collected on the day of
diagnosis in order to calculate IDSA severity scores.
The occurrence of ICU admission, colectomy, or death within 30 days was recorded
and in some cases, physicians conducted chart review to determine whether the
complication was attributable to the CDI.
In Chapter 4, we used results from surveys of learners who participated in the
Girls Who Code club and Carpentries workshop where we piloted our new
curricula, which allowed us to measure the success of our teaching approaches.
The data are described in further detail in each of their respective chapters.

