<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Improving machine learning models for gut microbiome classification problems and building resources to democratize data science - 3&nbsp; Democratizing data science with open curricula and user-friendly software tools</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/cdiff.html" rel="next">
<link href="../chapters/optifit.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<meta name="twitter:title" content="Improving machine learning models for gut microbiome classification problems and building resources to democratize data science - 3&nbsp; Democratizing data science with open curricula and user-friendly software tools">
<meta name="twitter:description" content="TODO">
<meta name="twitter:image" content="http://sovacool.dev/phd-diss/papers/girls-who-code/paper/lesson-modules.png">
<meta name="twitter:image-height" content="1423">
<meta name="twitter:image-width" content="2530">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/data-science.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Democratizing data science with open curricula and user-friendly software tools</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Democratizing data science with open curricula and user-friendly software tools</span></h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  <div>
    <div class="abstract">
      <div class="abstract-title">Abstract</div>
      <p>TODO</p>
    </div>
  </div>
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Improving machine learning models for gut microbiome classification problems and building resources to democratize data science</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/kelly-sovacool/phd-diss" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <a href="../dissertation_KLS.pdf" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/optifit.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">OptiFit: an improved method for fitting amplicon sequences to existing OTUs</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/data-science.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Democratizing data science with open curricula and user-friendly software tools</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/cdiff.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Predicting <em>C. difficile</em> infection severity from the taxonomic composition of the gut microbiome</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/discussion.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Discussion</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#preamble" id="toc-preamble" class="nav-link active" data-scroll-target="#preamble"><span class="header-section-number">3.1</span> Preamble</a></li>
  <li><a href="#teaching-python-for-data-science-collaborative-development-of-a-modular-and-interactive-curriculum" id="toc-teaching-python-for-data-science-collaborative-development-of-a-modular-and-interactive-curriculum" class="nav-link" data-scroll-target="#teaching-python-for-data-science-collaborative-development-of-a-modular-and-interactive-curriculum"><span class="header-section-number">3.2</span> Teaching Python for Data Science: Collaborative development of a modular and interactive curriculum</a>
  <ul class="collapse">
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">3.2.1</span> Summary</a></li>
  <li><a href="#statement-of-need" id="toc-statement-of-need" class="nav-link" data-scroll-target="#statement-of-need"><span class="header-section-number">3.2.2</span> Statement of Need</a></li>
  <li><a href="#collaborative-curriculum-development" id="toc-collaborative-curriculum-development" class="nav-link" data-scroll-target="#collaborative-curriculum-development"><span class="header-section-number">3.2.3</span> Collaborative Curriculum Development</a></li>
  <li><a href="#curriculum" id="toc-curriculum" class="nav-link" data-scroll-target="#curriculum"><span class="header-section-number">3.2.4</span> Curriculum</a></li>
  <li><a href="#instructional-design" id="toc-instructional-design" class="nav-link" data-scroll-target="#instructional-design"><span class="header-section-number">3.2.5</span> Instructional Design</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements"><span class="header-section-number">3.2.6</span> Acknowledgements</a></li>
  <li><a href="#funding" id="toc-funding" class="nav-link" data-scroll-target="#funding"><span class="header-section-number">3.2.7</span> Funding</a></li>
  <li><a href="#author-contributions" id="toc-author-contributions" class="nav-link" data-scroll-target="#author-contributions"><span class="header-section-number">3.2.8</span> Author Contributions</a></li>
  <li><a href="#conflicts-of-interest" id="toc-conflicts-of-interest" class="nav-link" data-scroll-target="#conflicts-of-interest"><span class="header-section-number">3.2.9</span> Conflicts of Interest</a></li>
  </ul></li>
  <li><a href="#developing-and-deploying-an-integrated-workshop-curriculum-teaching-computational-skills-for-reproducible-research" id="toc-developing-and-deploying-an-integrated-workshop-curriculum-teaching-computational-skills-for-reproducible-research" class="nav-link" data-scroll-target="#developing-and-deploying-an-integrated-workshop-curriculum-teaching-computational-skills-for-reproducible-research"><span class="header-section-number">3.3</span> Developing and deploying an integrated workshop curriculum teaching computational skills for reproducible research</a>
  <ul class="collapse">
  <li><a href="#summary-1" id="toc-summary-1" class="nav-link" data-scroll-target="#summary-1"><span class="header-section-number">3.3.1</span> Summary</a></li>
  <li><a href="#statement-of-need-1" id="toc-statement-of-need-1" class="nav-link" data-scroll-target="#statement-of-need-1"><span class="header-section-number">3.3.2</span> Statement of Need</a></li>
  <li><a href="#collaborative-curriculum-development-1" id="toc-collaborative-curriculum-development-1" class="nav-link" data-scroll-target="#collaborative-curriculum-development-1"><span class="header-section-number">3.3.3</span> Collaborative Curriculum Development</a></li>
  <li><a href="#curriculum-1" id="toc-curriculum-1" class="nav-link" data-scroll-target="#curriculum-1"><span class="header-section-number">3.3.4</span> Curriculum</a></li>
  <li><a href="#acknowledgements-1" id="toc-acknowledgements-1" class="nav-link" data-scroll-target="#acknowledgements-1"><span class="header-section-number">3.3.5</span> Acknowledgements</a></li>
  <li><a href="#funding-1" id="toc-funding-1" class="nav-link" data-scroll-target="#funding-1"><span class="header-section-number">3.3.6</span> Funding</a></li>
  <li><a href="#author-contributions-1" id="toc-author-contributions-1" class="nav-link" data-scroll-target="#author-contributions-1"><span class="header-section-number">3.3.7</span> Author Contributions</a></li>
  <li><a href="#conflicts-of-interest-1" id="toc-conflicts-of-interest-1" class="nav-link" data-scroll-target="#conflicts-of-interest-1"><span class="header-section-number">3.3.8</span> Conflicts of Interest</a></li>
  </ul></li>
  <li><a href="#mikropml-user-friendly-r-package-for-supervised-machine-learning-pipelines" id="toc-mikropml-user-friendly-r-package-for-supervised-machine-learning-pipelines" class="nav-link" data-scroll-target="#mikropml-user-friendly-r-package-for-supervised-machine-learning-pipelines"><span class="header-section-number">3.4</span> mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines</a>
  <ul class="collapse">
  <li><a href="#summary-2" id="toc-summary-2" class="nav-link" data-scroll-target="#summary-2"><span class="header-section-number">3.4.1</span> Summary</a></li>
  <li><a href="#statement-of-need-2" id="toc-statement-of-need-2" class="nav-link" data-scroll-target="#statement-of-need-2"><span class="header-section-number">3.4.2</span> Statement of need</a></li>
  <li><a href="#mikropml-package" id="toc-mikropml-package" class="nav-link" data-scroll-target="#mikropml-package"><span class="header-section-number">3.4.3</span> mikropml package</a></li>
  <li><a href="#acknowledgments" id="toc-acknowledgments" class="nav-link" data-scroll-target="#acknowledgments"><span class="header-section-number">3.4.4</span> Acknowledgments</a></li>
  <li><a href="#funding-2" id="toc-funding-2" class="nav-link" data-scroll-target="#funding-2"><span class="header-section-number">3.4.5</span> Funding</a></li>
  <li><a href="#author-contributions-2" id="toc-author-contributions-2" class="nav-link" data-scroll-target="#author-contributions-2"><span class="header-section-number">3.4.6</span> Author contributions</a></li>
  <li><a href="#conflicts-of-interest-2" id="toc-conflicts-of-interest-2" class="nav-link" data-scroll-target="#conflicts-of-interest-2"><span class="header-section-number">3.4.7</span> Conflicts of interest</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="preamble" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="preamble"><span class="header-section-number">3.1</span> Preamble</h2>
<p>TODO introduce 3 papers and explain how they tie together</p>
<!-- GIRLS WHO CODE -->
</section>
<section id="teaching-python-for-data-science-collaborative-development-of-a-modular-and-interactive-curriculum" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="teaching-python-for-data-science-collaborative-development-of-a-modular-and-interactive-curriculum"><span class="header-section-number">3.2</span> Teaching Python for Data Science: Collaborative development of a modular and interactive curriculum</h2>
<p>Co-authors of the Girls Who Code paper include: Marlena Duda<sup>*</sup>, Kelly L. Sovacool<sup>*</sup>, Negar Farzaneh, Vy Kim Nguyen, Sarah E. Haynes, Hayley Falk, Katherine L. Furman, Logan A. Walker, Rucheng Diao, Morgan Oneka, Audrey C. Drotos, Alana Woloshin, Gabrielle A. Dotson, April Kriebel, Lucy Meng, Stephanie N. Thiede, Zena Lapp, and Brooke N. Wolford.</p>
<p><sup>*</sup>Indicates co-first author</p>
<section id="summary" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="summary"><span class="header-section-number">3.2.1</span> Summary</h3>
<p>We are bioinformatics trainees at the University of Michigan who started a local chapter of Girls Who Code to provide a fun and supportive environment for high school women to learn the power of coding. Our goal was to cover basic coding topics and data science concepts through live coding and hands-on practice. However, we could not find a resource that exactly met our needs. Therefore, over the past three years, we have developed a curriculum and instructional format using Jupyter notebooks to effectively teach introductory Python for data science. This method, inspired by The Carpentries organization, uses bite-sized lessons followed by independent practice time to reinforce coding concepts, and culminates in a data science capstone project using real-world data. We believe our open curriculum is a valuable resource to the wider education community and hope that educators will use and improve our lessons, practice problems, and teaching best practices. Anyone can contribute to our Open Educational Resources <a href="https://github.com/GWC-DCMB">on GitHub</a>.</p>
</section>
<section id="statement-of-need" class="level3" data-number="3.2.2">
<h3 data-number="3.2.2" class="anchored" data-anchor-id="statement-of-need"><span class="header-section-number">3.2.2</span> Statement of Need</h3>
<p>As women bioinformatics trainees at the University of Michigan (U-M), we experience the gender gap in our field first-hand. During the 1974-1975 academic year, women achieved 18.9% of total Bachelor’s degrees in computer and information sciences in the US <span class="citation" data-cites="nces_digest_2012">(<a href="#ref-nces_digest_2012" role="doc-biblioref">National Center for Education Statistics 2012</a>)</span>. By 1983-1984 this peaked at 37.1%, but fell to 17.6% by 2010-2011. We also see this national trend in the training of the next generation of Bioinformaticians at Michigan Medicine. Since accepting its first students in 2001, the U-M Bioinformatics Graduate Program has graduated 66 male and 22 female doctorates as of 2019. This disparity begins at the applicant level; during 2016-2019 the average percentage of females applying directly to the Bioinformatics PhD program was 35.2%, and the average percentage of female applicants listing Bioinformatics as first or second choice in the Program in Biomedical Sciences, U-M’s biomedical PhD umbrella program was 41%.</p>
<p>Previous research on women’s educational experiences in science, technology, engineering, and mathematics (STEM) have produced various explanations for persistent gender disparities <span class="citation" data-cites="benbow_gender_2016">(<a href="#ref-benbow_gender_2016" role="doc-biblioref">Benbow and Vivyan 2016</a>)</span>. One explanation is that women often experience stereotype threats that negatively influence their math and science performance and deter them from pursuing STEM as a career <span class="citation" data-cites="hill_why_2010">(<a href="#ref-hill_why_2010" role="doc-biblioref">Hill, Corbett, and St. Rose 2010</a>)</span>. The majority of our organization’s founding graduate students (all women) began coding in our undergraduate careers or later. We wanted to provide a safe environment for local high school women to develop confidence in themselves and their computational skills before college, and be exposed to successful women role models in STEM to counter negative stereotypes.</p>
<p>Girls Who Code, a national organization whose mission is to close the gender gap in technology <span class="citation" data-cites="saujani_girls_2015">(<a href="#ref-saujani_girls_2015" role="doc-biblioref">Saujani 2015</a>)</span>, was founded in 2012. Because of our personal experiences and the paucity of women in our field <span class="citation" data-cites="nces_digest_2012 bonham_women_2017">(<a href="#ref-nces_digest_2012" role="doc-biblioref">National Center for Education Statistics 2012</a>; <a href="#ref-bonham_women_2017" role="doc-biblioref">Bonham and Stefan 2017</a>)</span>, we began a Girls Who Code student organization at the University of Michigan in 2017. For the past four academic years we have registered annually as a recognized Girls Who Code Club because the national organization provides name recognition, curriculum resources, guidance for a Capstone Impact Project, and a framework for launching a coding club. Participants in the Club attend weekly meetings at the University of Michigan (when the club is run in person rather than virtually), and are thus largely high school women from the Ann Arbor area. In 2019 we launched our own summer program, the Data Science Summer Experience. When held in person, the Summer Experience is hosted in Detroit to provide the opportunity for high school women outside of Ann Arbor to learn coding skills in an inclusive environment.</p>
<p>The national Girls Who Code organization provides a curriculum that teaches website and application development through programming languages like HTML and Java; however, our biomedical science graduate students generally have limited experience with these languages and with web development. In contrast, many of us have extensive experience performing data science using the Python programming language. Data Scientist was rated the #1 job in America by Glassdoor in 2016-2019, #3 in 2020, and #2 in 2021 <span class="citation" data-cites="stansell_breaking_2019">(<a href="#ref-stansell_breaking_2019" role="doc-biblioref">Stansell 2019</a>)</span>. Furthermore, Python is the most popular programming language according to the PYPL PopularitY of Programming Language Index <span class="citation" data-cites="noauthor_pypl_nodate">(<a href="#ref-noauthor_pypl_nodate" role="doc-biblioref"><span>“<span>PYPL PopularitY</span> of <span>Programming Language</span> Index,”</span> n.d.</a>)</span>. Therefore, we believe career exploration in data science using the Python programming language will optimally prepare our learners for careers that provide financial stability and upward economic mobility. By leveraging the data science expertise of our Club facilitators (hereafter termed instructors), we created a specialized curriculum focused on computational data science in the Python programming language.</p>
<p>Girls Who Code encourages participants to learn programming skills while working on an Impact Project website or application throughout the Club <span class="citation" data-cites="hq_gallery_2021">(<a href="#ref-hq_gallery_2021" role="doc-biblioref">Girls Who Code HQ 2021</a>)</span>. We created an open source Data Science curriculum that teaches the requisite Python and statistics skills to complete a Capstone Project, where learners explore, analyze, and present a data set of their choosing. Using this curriculum, we employ participatory live coding, where learners type and run code along with the instructor in real time. Using paired activities, our curriculum follows the “I do, we do, you do” didactic paradigm <span class="citation" data-cites="fisher_better_2013">(<a href="#ref-fisher_better_2013" role="doc-biblioref">D. Fisher and Frey 2013</a>)</span>. We provide open source resources for both in-person and virtual versions of our curriculum, including videos corresponding to each lesson. While we developed this curriculum for our Girls Who Code Club and Summer Experience, we believe that it can be widely used for teaching introductory coding for data science.</p>
</section>
<section id="collaborative-curriculum-development" class="level3" data-number="3.2.3">
<h3 data-number="3.2.3" class="anchored" data-anchor-id="collaborative-curriculum-development"><span class="header-section-number">3.2.3</span> Collaborative Curriculum Development</h3>
<p>We assembled a team of volunteers involved in our club to develop a custom curriculum to teach introductory Python for data science. We chose the content based on what our learners would need to learn to complete a small data analysis project and communicate their findings to their peers. We divided the content by topic into Jupyter notebooks for each lesson, with each lesson taking approximately 15-20 minutes to teach via live coding. Every lesson has a corresponding practice notebook with additional exercises on the same content taught in the lesson, but using different data or variables. We used a similar development workflow as the U-M Carpentries curriculum <span class="citation" data-cites="lapp_developing_2021">(<a href="#ref-lapp_developing_2021" role="doc-biblioref">Lapp et al. 2021</a>)</span>. Briefly, we hosted the curriculum notebooks in a public GitHub repository to facilitate collaborative development and peer review using pull requests. In the initial curriculum drafting phase, developers were assigned lesson and practice notebooks to write. Once the draft of a lesson was completed, the writer opened a pull request and asked for review from a different developer. The reviewer then provided feedback and approved the pull request to be merged into the main branch after the writer made any requested changes. This way, more than one person viewed each notebook before it could be incorporated into the public curriculum, which reduced mistakes and ensured higher quality content. While teaching from the curriculum at the first Data Science Summer Experience, instructors took notes on their experience and made revisions afterward. Maintainers continue to monitor the repository and resolve issues as they arise.</p>
<p>Following the onset of the COVID-19 pandemic, we quickly pivoted our club to a virtual format. In preparation for the 2020 Summer Experience, we switched to a flipped classroom style following feedback from our club participants that it was too difficult to follow along live coding via Zoom (see <a href="#instructional-design">Instructional Design</a>).</p>
</section>
<section id="curriculum" class="level3" data-number="3.2.4">
<h3 data-number="3.2.4" class="anchored" data-anchor-id="curriculum"><span class="header-section-number">3.2.4</span> Curriculum</h3>
<p>Our curriculum was designed for high school students with no prior coding experience who are interested in learning Python programming for data science. However, this course material would be useful for anyone interested in teaching or learning basic programming for data analysis.</p>
<section id="learning-objectives" class="level4" data-number="3.2.4.1">
<h4 data-number="3.2.4.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">3.2.4.1</span> Learning Objectives</h4>
<p>The learning objectives of this curriculum are:</p>
<ol type="1">
<li>Write code in Python with correct syntax and following best practices.</li>
<li>Implement fundamental programming concepts when presented with a programmatic problem set.</li>
<li>Apply data analysis to real world data to answer scientific questions.</li>
<li>Create informative summary statistics and data visualizations in Python.</li>
</ol>
<p>These skills provide a solid foundation for basic data analysis in Python. Participation in our program exposes learners to the many ways coding and data science can be impactful across many disciplines.</p>
</section>
<section id="course-content" class="level4" data-number="3.2.4.2">
<h4 data-number="3.2.4.2" class="anchored" data-anchor-id="course-content"><span class="header-section-number">3.2.4.2</span> Course Content</h4>
<p>Our curriculum design consists of 27 lessons broken up into 5 modules that cover Jupyter notebook setup, Python coding fundamentals, use of essential data science packages including pandas and numpy, basic statistical analyses, and plotting using seaborn and matplotlib (Figure 1) <span class="citation" data-cites="harris_array_2020 waskom_seaborn_2021 hunter_matplotlib_2007">(<a href="#ref-harris_array_2020" role="doc-biblioref">Harris et al. 2020</a>; <a href="#ref-waskom_seaborn_2021" role="doc-biblioref">Waskom 2021</a>; <a href="#ref-hunter_matplotlib_2007" role="doc-biblioref">Hunter 2007</a>)</span>. Each lesson consists of a lesson notebook and a practice notebook containing similar exercises for the learner to complete on their own following the lesson.</p>
<p><img src="../papers/girls-who-code/paper/lesson-modules.png" class="img-fluid" alt="Our lesson modules. All Jupyter notebooks are available on GitHub (https://github.com/GWC-DCMB/curriculum-notebooks)."><code>r sovatools::get_image_dims(here::here('papers/girls-who-code/paper/lesson-modules.png'))</code></p>
<p>Each lesson builds on those before it, beginning with relevant content reminders from the previous lessons and ending with a concise summary of the skills presented within. As they progress through the curriculum, the learners begin simultaneously working on a data science project using a real world dataset of their choosing. While more time is dedicated to lessons early in the program, the formal curriculum tapers off until the learners are solely applying their skills to the data science project. Through this Capstone Project, learners gain practical experience with each skill as they learn it in the lessons; including importing and cleaning data, data visualization, and basic statistical analyses.</p>
</section>
</section>
<section id="instructional-design" class="level3" data-number="3.2.5">
<h3 data-number="3.2.5" class="anchored" data-anchor-id="instructional-design"><span class="header-section-number">3.2.5</span> Instructional Design</h3>
<p>We modeled our instructional design in the style of Software Carpentry <span class="citation" data-cites="wilson_software_2016">(<a href="#ref-wilson_software_2016" role="doc-biblioref">Wilson 2016</a>)</span>.</p>
<ol type="1">
<li>Each lesson begins with a recapping of the relevant core skills presented in the previous lessons.</li>
<li>All lessons are designed to be taught via 15-minute live-coding sessions. This method is used by <a href="https://carpentries.org/">The Carpentries</a> and is demonstrated to be an effective method that engages learners <span class="citation" data-cites="wilson_software_2016 nederbragt_ten_2020">(<a href="#ref-wilson_software_2016" role="doc-biblioref">Wilson 2016</a>; <a href="#ref-nederbragt_ten_2020" role="doc-biblioref">Nederbragt et al. 2020</a>)</span> since learners must actively engage with the material and deal with errors and bugs as they arise.</li>
<li>Each lesson ends with a summary of core skills presented within the material.</li>
<li>Each short lesson is also accompanied by a subsequent 10-minute independent practice, providing further opportunity for practical experience implementing the coding skill at hand and testing learners’ understanding of the content.</li>
</ol>
<p>To better facilitate virtual instruction during the COVID-19 pandemic, we switched to a flipped classroom. Prior to meeting, learners watch videos of instructors explaining the material through “live” coding and code along in the lesson notebook with while watching the video. Each video shows the Jupyter notebook alongside the instructor themselves teaching. Learners then complete a practice notebook corresponding to the lesson. During the virtual meeting time, instructors answer questions and review the core concepts in the practice exercises. This virtual format is especially beneficial because it 1) allows learners to learn at their own pace, and 2) enables dissemination of our curriculum to a wider audience interested in learning introductory Python programming for data science.</p>
<p>For both in-person and virtual instruction, once learners have completed the Fundamentals module and reach the Data Science Essentials module they begin simultaneous work on their data science projects. Projects are completed in a pair programming style, where partners take turns assuming the “driver” (i.e.&nbsp;the typer) and “navigator” (i.e.&nbsp;the helper) roles <span class="citation" data-cites="hannay_effectiveness_2009">(<a href="#ref-hannay_effectiveness_2009" role="doc-biblioref">Hannay et al. 2009</a>)</span>. Switching off in this way helps both partners assume equal responsibility for the project workload, but more importantly it enables improved knowledge transfer through peer-to-peer learning. The culmination of the project is a presentation to peers, instructors, and family members. Through this process learners gain hands-on experience coding, cleaning data, performing statistical analyses, creating informative data visualizations, and communicating their results to others.</p>
<p>In addition to our coding curriculum, another key component of our programming is hosting women guest speakers from diverse fields across academia and industry. Our guest speakers come to discuss the journey they have taken to their career paths as well as how they utilize programming and data science in their jobs. These varied perspectives are extremely valuable to our learners as they provide several practical examples of programming careers in the real world, and expose them to successful women in STEM.</p>
<section id="experience-of-use" class="level4" data-number="3.2.5.1">
<h4 data-number="3.2.5.1" class="anchored" data-anchor-id="experience-of-use"><span class="header-section-number">3.2.5.1</span> Experience of Use</h4>
<p>We have used this curriculum to teach the Data Science Summer Experience and Girls Who Code Club in person in 2019 and virtually in 2020-2021. For both in-person and virtual instances, we had several instructors present at each session to answer questions and help learners debug. Furthermore, one or two instructors were assigned to each project group to help learners define data analysis questions, develop and execute a data analysis plan, visualize and communicate their findings, and troubleshoot coding problems. Projects have ranged from investigating exoplanets to studying the genomics of psoriasis.</p>
<p>We credit the success of our curriculum not only to the skill of the instructors, but also to the way we organized and executed the lessons and project:</p>
<ol type="1">
<li>The instructors and learners used <a href="https://colab.research.google.com/">Google Colaboratory (Colab)</a> to write and execute code in Jupyter notebooks. We chose this option because learners do not have to install any programs to use Google Colab and can easily open and edit the Jupyter notebooks from GitHub. When meeting in person, most learners use Google Chromebooks which have limited programming capabilities, but easy use of a web browser.</li>
<li>Assigning instructors to groups allowed learners to build a more personal connection with their instructors, making them feel more comfortable asking questions.</li>
<li>Group projects were performed using pair programming to allow learners to collaborate and learn from each other.</li>
<li>We used the “sticky note” system from The Carpentries by which learners can ask for help by putting up a colored sticky note (or a Zoom emoji in the case of virtual meetings) <span class="citation" data-cites="becker_responding_2016">(<a href="#ref-becker_responding_2016" role="doc-biblioref">Becker 2016</a>)</span>.</li>
<li>We exposed the learners to different aspects of data science by bringing in women guest speakers from academics and industry. This allowed them to better put what they were learning into context, think about how they might use the skills they were learning in potential future careers, and exposed them to successful women in STEM.</li>
</ol>
<section id="learner-experiences" class="level5" data-number="3.2.5.1.1">
<h5 data-number="3.2.5.1.1" class="anchored" data-anchor-id="learner-experiences"><span class="header-section-number">3.2.5.1.1</span> Learner experiences</h5>
<p>We surveyed learners anonymously after each Club and Summer Experience and found that most felt that their skills in Python programming, problem solving, critical thinking, and collaboration had improved (Figure 2). Furthermore, on a 10 question skills assessment during the 2019-2020 instance of the Club, the average increase in correct answers between the first meeting and the last meeting was 4.2 with a standard deviation of 2.8 (N=5 respondents). We also surveyed Club and Summer Experience alumni and found that 75% (N=20) want to pursue a STEM career. 62% (N=21) are still coding. On a 5-point scale from ‘Strongly Disagree’ to ‘Strongly Agree,’ the average answer for ‘My participation in GWC impacted my career aspirations’ is 4 (s.d.=0.9), with 4.5 (s.d.=0.6) for ‘Participating in GWC made me feel more confident in analyzing data’ and 3.9 (s.d.=1) for ‘Participating in GWC made me more confident in myself.’</p>
<p><img src="../papers/girls-who-code/paper/skill_plots.png" class="img-fluid" alt="Post-survey responses. Learners were asked if they felt that their skills in Python programming, problem solving, critical thinking, and collaboration had improved."><code>r sovatools::get_image_dims(here::here('papers/girls-who-code/paper/skill_plots.png'))</code></p>
<p>Overwhelmingly, learners’ favorite parts of the program are the guest speakers and the project. These aspects of our curriculum expose them to new fields and allow them to apply their newfound coding skills to asking an interesting question. A 2021 Club learner shared, “I plan to go to college for Computer Science and get a robotics minor when my college offers it. GWC has inspired me to consider pursuing a Masters or PhD in CS as well as take some electives in Data Science.” Five of our 86 alumni have gone on to perform research with U-M faculty members, with one presenting her work at an international conference. In fact, about a third of participants claim that they are now more interested in pursuing a career in computer or data science compared to before their Girls Who Code experience.</p>
</section>
</section>
</section>
<section id="acknowledgements" class="level3" data-number="3.2.6">
<h3 data-number="3.2.6" class="anchored" data-anchor-id="acknowledgements"><span class="header-section-number">3.2.6</span> Acknowledgements</h3>
<p>We would like to acknowledge our faculty co-sponsors Maureen Sartor &amp; Cristina Mitrea. We appreciate the continued support of U-M DCMB staff and faculty including Julia Eussen, Mary Freer, Linda Peasley, Jane Wiesner, Brian Athey, and Margit Burmeister. We are grateful for the resources provided by the national Girls Who Code organization.</p>
<p>Our programming is made possible by the dedication of past and present Executive Committee members, Club and Summer Experience Facilitators, and Capstone Project mentors including Shweta Ramdas, Alex Weber, Arushi Varshney, Sophie Hoffman, Hojae Lee, Ruma Deb, Saige Rutherford, Michelle McNulty, Bailey Peck, Chloe Whicker, Carolina Rojas Ramirez, Verity Sturm, Zoe Drasner, Sarah Latto, Emily Roberts, Angel Chu, Vivek Rai, Hillary Miller, Ashton Baker, Murchtricia Jones, Lauren Jepsen, Aubrey Annis, Awanti Sambarey, Mengtong Hu, Maribel Okiye, Yingxiao Zhang, and Neslihan Bisgin.</p>
<p>We are grateful for the funding, assistance, and other support provided to our student organization from the following sponsors: the U-M Department of Computational Medicine and Bioinformatics, the U-M Department of Biostatistics, the U-M Department of Statistics, the U-M Office of Graduate and Postdoctoral Studies, the U-M Endowment in Basic Sciences, the U-M Detroit Center, the U-M Life Sciences Institute, the U-M Office of Research, the Michigan Council of Women in Technology Foundation, DELL Technologies, Cisco Systems, Zingerman’s Delicatessen, the Girls Who Code Support Fund, and anonymous donations from Giving Blue Day 2019.</p>
<p>We also thank the learners who have participated in our Club and Summer Experience events.</p>
</section>
<section id="funding" class="level3" data-number="3.2.7">
<h3 data-number="3.2.7" class="anchored" data-anchor-id="funding"><span class="header-section-number">3.2.7</span> Funding</h3>
<p>MD, ACD, ZL, and BNW received support from the National Science Foundation Graduate Research Fellowship Program under Grant No.&nbsp;DGE 1256260. Any opinions, findings, and conclusions or recommendations expressed in this material are those of the authors and do not necessarily reflect the views of the National Science Foundation.</p>
<p>MD, KLS, NF, and VKN received support from the NIH Training Program in Bioinformatics (T32 GM070449). NF was supported by the National Institute of Health (NIH) Ruth L. Kirschstein National Research Service Award (NRSA) Individual Predoctoral Fellowship Program (F31 LM012946-01). VKN was supported by a NIH Research Project Grant on Breast Cancer Disparities (RO1-ES028802) and the CDC through the National Institute for Occupational Safety and Health (NIOSH) Pilot Project Research Training Program (T42-OH008455). KLF received support from The University of Michigan NIDA Training Program in Neuroscience (T32-DA7281) and from the NIH Early Stage Training in the Neurosciences Training Grant (T32-NS076401). MO received support from the Advanced Proteome Informatics of Cancer Training Grant (T32 CA140044). SNT was supported by the Molecular Mechanisms in Microbial Pathogenesis training grant (NIH T32 AI007528). ZL and BNW received support from the NIH Training Program in Genomic Science (T32-HG000040-22).</p>
</section>
<section id="author-contributions" class="level3" data-number="3.2.8">
<h3 data-number="3.2.8" class="anchored" data-anchor-id="author-contributions"><span class="header-section-number">3.2.8</span> Author Contributions</h3>
<p>MD, KLS, ZL, and BNW wrote the initial draft of the manuscript. All authors contributed to the curriculum and reviewed the manuscript.</p>
</section>
<section id="conflicts-of-interest" class="level3" data-number="3.2.9">
<h3 data-number="3.2.9" class="anchored" data-anchor-id="conflicts-of-interest"><span class="header-section-number">3.2.9</span> Conflicts of Interest</h3>
<p>None.</p>
<!-- SOFTWARE CARPENTRY -->
</section>
</section>
<section id="developing-and-deploying-an-integrated-workshop-curriculum-teaching-computational-skills-for-reproducible-research" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="developing-and-deploying-an-integrated-workshop-curriculum-teaching-computational-skills-for-reproducible-research"><span class="header-section-number">3.3</span> Developing and deploying an integrated workshop curriculum teaching computational skills for reproducible research</h2>
<p>Co-authors of the Software Carpentry paper include: Zena Lapp<sup>*</sup>, Kelly L. Sovacool<sup>*</sup>, Nick Lesniak, Dana King, Catherine Barnier, Matthew Flickinger, Jule Krüger, Courtney R. Armour, Maya M. Lapp, Jason Tallant, Rucheng Diao, Morgan Oneka, Sarah Tomkovich, Jacqueline Moltzau Anderson, Sarah K. Lucas, and Patrick D. Schloss.</p>
<p><sup>*</sup>Indicates co-first author</p>
<section id="summary-1" class="level3" data-number="3.3.1">
<h3 data-number="3.3.1" class="anchored" data-anchor-id="summary-1"><span class="header-section-number">3.3.1</span> Summary</h3>
<p>Inspired by well-established material and pedagogy provided by The Carpentries <span class="citation" data-cites="wilson_software_2016">(<a href="#ref-wilson_software_2016" role="doc-biblioref">Wilson 2016</a>)</span>, we developed a two-day workshop curriculum that teaches introductory R programming for managing, analyzing, plotting and reporting data using packages from the tidyverse <span class="citation" data-cites="wickham_welcome_2019">(<a href="#ref-wickham_welcome_2019" role="doc-biblioref">Wickham et al. 2019</a>)</span>, the Unix shell, version control with git, and GitHub. While the official Software Carpentry curriculum is comprehensive, we found that it contains too much content for a two-day workshop. We also felt that the independent nature of the lessons left learners confused about how to integrate the newly acquired programming skills in their own work. Thus, we developed <a href="https://umcarpentries.org/intro-curriculum-r/">a new curriculum</a> that aims to teach novices how to implement reproducible research principles in their own data analysis. The curriculum integrates live coding lessons with individual-level and group-based practice exercises, and also serves as a succinct resource that learners can reference both during and after the workshop. Moreover, it lowers the entry barrier for new instructors as they do not have to develop their own teaching materials or sift through extensive content. We developed this curriculum during a two-day sprint, successfully used it to host a two-day virtual workshop with almost 40 participants, and updated the material based on instructor and learner feedback. We hope that our new curriculum will prove useful to future instructors interested in teaching workshops with similar learning objectives.</p>
</section>
<section id="statement-of-need-1" class="level3" data-number="3.3.2">
<h3 data-number="3.3.2" class="anchored" data-anchor-id="statement-of-need-1"><span class="header-section-number">3.3.2</span> Statement of Need</h3>
<p>For the past five years, the University of Michigan instance of The Carpentries has taught workshops using versions of curriculum originally created by The Carpentries organization. In that time, our instructors found several advantages and disadvantages to using the original Software Carpentry curriculum. Some of the advantages were that any programming language lesson (e.g., R or Python) could be paired with lessons on the Unix shell and version control, lessons had been refined by many contributors over the years and taught at workshops around the world, and the instructional design demonstrated good pedagogy for teaching novice data science practitioners. However, The Carpentries materials have evolved from lesson plans to reference materials, and thus there was too much content for the time available during a two-day workshop. As a result, workshops taught with this material were inconsistent depending on who was teaching, and new instructors faced an overwhelming amount of work to prepare for their first workshop. Furthermore, the modular nature of the curriculum meant that each lesson was independent from the others, so it was not apparent to learners how all of the skills could be integrated for the purpose of a reproducible research project.</p>
<p>Given these constraints, we sought to create a new curriculum that would allow us to teach computational skills in an integrated manner, demonstrate the reproducible research workflows we use in our own work, deliver an appropriate and consistent amount of content, and reduce the burden for new instructors to get involved, all while maintaining the same inclusive pedagogy that has been refined by The Carpentries organization.</p>
</section>
<section id="collaborative-curriculum-development-1" class="level3" data-number="3.3.3">
<h3 data-number="3.3.3" class="anchored" data-anchor-id="collaborative-curriculum-development-1"><span class="header-section-number">3.3.3</span> Collaborative Curriculum Development</h3>
<p>We drew on the expertise of The Carpentries community at the University of Michigan to develop a custom curriculum that would meet our goals (Figure 1). To start, we organized a two-day sprint, where members of our community worked collaboratively to create an initial draft of the content. During the sprint, we met virtually to discuss our goals, then broke up into teams to work on individual lessons before coming back together to review our progress. We hosted the curriculum in a public GitHub repository (https://github.com/umcarpentries/intro-curriculum-r) to facilitate collaborative work and peer review using issues, branches, and pull requests. Under this model, a team member created or edited content in a new branch to resolve an issue, then created a pull request and asked for review from another team member, who finally merged the changes into the default branch. GitHub pages automatically uses the default branch to build a website that allows us to host the polished curriculum (https://umcarpentries.org/intro-curriculum-r/). Our collaborative model ensured that at least two pairs of eyes viewed any changes before they could be included in the curriculum. This strategy helped us reduce mistakes and create better quality content.</p>
<p><img src="../papers/software-carpentry/paper/development-framework.png" class="img-fluid" alt="The curriculum development framework. Created with BioRender.com."><code>r sovatools::get_image_dims(here::here('papers/software-carpentry/paper/development-framework.png'))</code></p>
<p>Following the sprint, contributors finalized edits and continued to review each others’ pull requests to complete the alpha version of our curriculum. Next, we hosted a workshop for instructors to pilot the curriculum. We collected feedback from the learners and instructors at the end of the pilot workshop and then held a smaller half-day sprint to revise the curriculum based on the feedback. Currently, our community members are continuously able to create issues, make edits, and review pull requests to keep refining the curriculum for future use. We are planning more workshops with new instructors who were not involved in the original curriculum development to gather their feedback.</p>
</section>
<section id="curriculum-1" class="level3" data-number="3.3.4">
<h3 data-number="3.3.4" class="anchored" data-anchor-id="curriculum-1"><span class="header-section-number">3.3.4</span> Curriculum</h3>
<p>Our curriculum is tailored to people with no prior coding experience who want to learn how to use R programming for data analysis, visualization and the reporting of results (Figure 2). Not only do we aim to teach our learners the basics of performing empirical data analysis, we also seek to provide a rigorous framework for adhering to reproducible research principles that enable researchers to easily share their empirical work with others.</p>
<p><img src="../papers/software-carpentry/paper/curriculum-overview.png" class="img-fluid" alt="Curriculum overview. Created with BioRender.com."><code>r sovatools::get_image_dims('papers/software-carpentry/paper/curriculum-overview.png')</code></p>
<section id="learning-objectives-1" class="level4" data-number="3.3.4.1">
<h4 data-number="3.3.4.1" class="anchored" data-anchor-id="learning-objectives-1"><span class="header-section-number">3.3.4.1</span> Learning Objectives</h4>
<p>The key learning objectives for our curriculum are:</p>
<ol type="1">
<li>Create clear and informative data visualizations in R, starting with messy data.</li>
<li>Perform version control using the Unix shell and git.</li>
<li>Create reproducible reports using R Markdown.</li>
<li>Share code with others on GitHub.</li>
</ol>
<p>We believe these skills provide learners with a solid foundation from which they can teach themselves any additional coding skills for future use.</p>
</section>
<section id="course-content-1" class="level4" data-number="3.3.4.2">
<h4 data-number="3.3.4.2" class="anchored" data-anchor-id="course-content-1"><span class="header-section-number">3.3.4.2</span> Course Content</h4>
<p>Our curriculum consists of nine modules that cover software setup, data analysis and visualization in R, version control, sharing code, and writing reproducible reports (see below for more details). The R programming lessons take a “tidyverse first” approach <span class="citation" data-cites="robinson_teach_2017">(<a href="#ref-robinson_teach_2017" role="doc-biblioref">Robinson 2017</a>)</span> to effectively and efficiently teach learners powerful tools for plotting and data analysis. We also set an overall goal for the workshop to make the content substantively interesting and relatable to a wide audience regardless of their original academic discipline or professional practice. Specifically, we task our learners with producing a fictitious report to the United Nations that examines the relationship between gross domestic product (GDP), life expectancy, and CO<sub>2</sub> emissions. The nine curriculum modules are:</p>
<ol start="0" type="1">
<li>Setup</li>
<li>Welcome</li>
<li>R for plotting (uses the tidyverse R packages <span class="citation" data-cites="wickham_welcome_2019">(<a href="#ref-wickham_welcome_2019" role="doc-biblioref">Wickham et al. 2019</a>)</span>)</li>
<li>The Unix shell</li>
<li>Git and GitHub</li>
<li>R for data analysis (uses the tidyverse R packages <span class="citation" data-cites="wickham_welcome_2019">(<a href="#ref-wickham_welcome_2019" role="doc-biblioref">Wickham et al. 2019</a>)</span>)</li>
<li>Writing reports in R Markdown (uses the rmarkdown R package <span class="citation" data-cites="xie_r_2018">(<a href="#ref-xie_r_2018" role="doc-biblioref">Xie, Allaire, and Grolemund 2018</a>)</span>)</li>
<li>Group practice exercises</li>
<li>Where to go from here</li>
</ol>
<p>Each lesson builds on the previous ones. The Unix shell, git, and GitHub are introduced using the files generated in the R for plotting lesson. The lesson content for subsequent modules is then intermittently committed and pushed to GitHub. The ‘Writing reports in R Markdown’ lesson combines all of the skills learned previously to produce a report that one could share with the United Nations. Next, learners put everything they have learned into practice by forming small groups and working on practice problems that cover the entire course content (<a href="https://umcarpentries.org/intro-curriculum-r/05-r-markdown/index.html#integrating-it-all-together-paired-exercise">“Integrating it all together: Paired exercise”</a>). The workshop completes with a short module recapping everything that the curriculum covered as well as offering suggestions on how learners can continue to get help and keep learning once the workshop ends.</p>
</section>
<section id="instructional-design-1" class="level4" data-number="3.3.4.3">
<h4 data-number="3.3.4.3" class="anchored" data-anchor-id="instructional-design-1"><span class="header-section-number">3.3.4.3</span> Instructional Design</h4>
<!-- teaching philosophy / pedagogy -->
<p>Our modules and teaching suggestions are developed in the style of <a href="https://software-carpentry.org/">Software Carpentry</a>:</p>
<ol type="1">
<li>Each module contains learning objectives at the beginning of each lesson and a summary of key points at the end.</li>
<li>The five core modules (2 to 6) are designed to be taught via live coding of the content to learners. This is a central feature of Carpentries lessons, and we believe it is a great way to learn how to program. It requires learners to follow along and encounter errors that they must debug along the way, fostering additional questions about the course content. It also leads to instructors making mistakes and then demonstrating how to deal with them in an ad hoc and iterative manner.</li>
<li>We incorporate formative assessments in the form of short practice exercises throughout each lesson such that learners can practice what they have learned, while instructors can gauge learner understanding of the material.</li>
<li>We use the “sticky note” system for formative assessment, where learners indicate their progress on exercises and request help by using different colored sticky notes <span class="citation" data-cites="becker_responding_2016 the_carpentries_live_2018">(<a href="#ref-becker_responding_2016" role="doc-biblioref">Becker 2016</a>; <a href="#ref-the_carpentries_live_2018" role="doc-biblioref">The Carpentries 2018a</a>)</span>. At virtual workshops, we use Zoom reaction icons as virtual sticky notes, with the red X reaction to ask for help and the green checkmark to indicate that an exercise was successfully completed.</li>
<li>We have several helpers attend each workshop to address learner questions and technical issues.</li>
</ol>
<p>We also incorporated a few additional key components into the curriculum:</p>
<ol type="1">
<li>Each lesson built off of previous lessons, with the goal of creating a final report that can be shared with others.</li>
<li>We structured the curriculum such that it could be taught through an in-person or virtual workshop. Virtual workshops are sometimes necessary, as during the COVID-19 pandemic, but are also useful to allow people from a variety of geographic locations to instruct and attend.</li>
<li>We not only required learners to install all software before the workshop (as The Carpentries also requires), but also asked them to run an example script that tests whether everything is installed correctly. To attend the workshop, learners were required to send screenshots of the script output to the workshop lead in advance. We withheld the login details for the workshop until we received the screenshot. This ensured that any installation issues could be addressed before the workshop began.</li>
<li>An extensive small group practice module towards the end of the workshop allowed learners to more independently practice the skills they have learned.</li>
<li>The workshop concluded with a recap of what was covered and resources available for learners to continue learning and getting help as their skills develop.</li>
</ol>
</section>
<section id="pilot-workshop" class="level4" data-number="3.3.4.4">
<h4 data-number="3.3.4.4" class="anchored" data-anchor-id="pilot-workshop"><span class="header-section-number">3.3.4.4</span> Pilot Workshop</h4>
<!-- experience of use -->
<p>We piloted our curriculum during a virtual two-day Software Carpentry workshop. In line with The Carpentries recommendations <span class="citation" data-cites="the_carpentries_carpentries_2018">(<a href="#ref-the_carpentries_carpentries_2018" role="doc-biblioref">The Carpentries 2018b</a>)</span>, we had four instructors and six helpers at the workshop to assist with learner questions and technical issues. We had thirty-nine learners of various skill levels from several different countries, all of whom provided very positive reviews of the workshop. To assess the effectiveness of the workshop, learners were asked to complete a pre- and post-workshop survey administered by the Carpentries. By the end of the workshop, learners on average felt more confident writing programs, using programming to work with data, overcoming problems while programming, and searching for answers to technical questions online (n = 14 survey respondents; see Figure 3). All attendees who filled out the post-workshop survey (n = 19) would recommend the workshop to others.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../papers/software-carpentry/paper/survey-results.png" style="width:100.0%;height:50.0%" class="figure-img"></p>
<figcaption class="figure-caption">Pre- and post-workshop survey results.</figcaption>
</figure>
</div>
<section id="virtual-workshop-reflection" class="level5" data-number="3.3.4.4.1">
<h5 data-number="3.3.4.4.1" class="anchored" data-anchor-id="virtual-workshop-reflection"><span class="header-section-number">3.3.4.4.1</span> Virtual Workshop Reflection</h5>
<p>We credit the success of our first virtual workshop in large part due to the curriculum structure and content, as well as the instructors and helpers involved. However, we also believe that the following helped make the workshop as smooth as possible:</p>
<ol type="1">
<li>We suggested that learners have Zoom and RStudio (or the Unix shell) open side-by-side on their computer to minimize toggling between different windows <span class="citation" data-cites="chen_online_2020">(<a href="#ref-chen_online_2020" role="doc-biblioref">D. Chen 2020</a>)</span>.</li>
<li>We used Slack for communication among instructors and helpers, as well as between helpers and learners. Learners asked questions in a group Slack channel where helpers could respond. This allowed us to address the vast majority of learner questions and bugs quickly, clearly, and efficiently without disrupting the lesson or moving the learner to a Zoom breakout room. Furthermore, Slack worked much better than the Zoom chat as questions could be answered in threads, were preserved and visible to all learners regardless of whether they were connected to Zoom at the time, and didn’t get lost as easily.<br>
</li>
<li>Whenever a learner needed more help than was possible on Slack, a helper and the learner entered a Zoom breakout room together to troubleshoot. However, we tried to minimize this option as much as possible to prevent the learner from missing content covered in the main room.</li>
</ol>
</section>
</section>
</section>
<section id="acknowledgements-1" class="level3" data-number="3.3.5">
<h3 data-number="3.3.5" class="anchored" data-anchor-id="acknowledgements-1"><span class="header-section-number">3.3.5</span> Acknowledgements</h3>
<p>We thank The Carpentries organization for providing instructor training, workshop protocols, and the open-source Software Carpentry curriculum upon which this curriculum is based. We also thank them for allowing us to use the pre- and post-workshop survey results in this manuscript. The Carpentries is a fiscally sponsored project of Community Initiatives, a registered 501(c)3 non-profit organisation based in California, USA.</p>
<p>We are grateful to Victoria Alden and Scott Martin for assisting us in organizing and advertising our pilot workshop. We thank Shelly Johnson for volunteering as a helper at the workshop and contributing to the setup instructions. We also thank Bennet Fauber for contributing to the setup instructions.</p>
<p>We thank the learners who participated in the workshop, provided feedback, and completed the surveys.</p>
</section>
<section id="funding-1" class="level3" data-number="3.3.6">
<h3 data-number="3.3.6" class="anchored" data-anchor-id="funding-1"><span class="header-section-number">3.3.6</span> Funding</h3>
<p>Salary support for PDS came from NIH grants R01CA215574 and U01AI124255. KLS received support from the NIH Training Program in Bioinformatics (T32 GM070449). ZL received support from the National Science Foundation Graduate Research Fellowship Program under Grant No.&nbsp;DGE 1256260. Any opinions, findings, and conclusions or recommendations expressed in this material are those of the authors and do not necessarily reflect the views of the National Science Foundation.</p>
</section>
<section id="author-contributions-1" class="level3" data-number="3.3.7">
<h3 data-number="3.3.7" class="anchored" data-anchor-id="author-contributions-1"><span class="header-section-number">3.3.7</span> Author Contributions</h3>
<p>ZL and KLS contributed equally. ZL is first among the co-first authors because KLS threatened to reject all pull requests where ZL put KLS first. :)</p>
<p>PDS supervised the project. ZL and KLS organized the initial sprint, led the development of the curriculum, and drafted the manuscript. ZL, KLS, JK, and MML instructed at the first pilot workshop while CRA, JMA, ST, SKL, and CB assisted learners. All authors contributed to the development of the curriculum.</p>
</section>
<section id="conflicts-of-interest-1" class="level3" data-number="3.3.8">
<h3 data-number="3.3.8" class="anchored" data-anchor-id="conflicts-of-interest-1"><span class="header-section-number">3.3.8</span> Conflicts of Interest</h3>
<p>None.</p>
<!-- MIKROPML -->
</section>
</section>
<section id="mikropml-user-friendly-r-package-for-supervised-machine-learning-pipelines" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="mikropml-user-friendly-r-package-for-supervised-machine-learning-pipelines"><span class="header-section-number">3.4</span> mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines</h2>
<p>Co-authors of the mikropml paper include: Begüm D. Topçuoğlu<sup>*</sup>, Zena Lapp<sup>*</sup>, Kelly L. Sovacool<sup>*</sup>, Evan Snitkin, Jenna Wiens, and Patrick D. Schloss.</p>
<p><sup>*</sup>Indicates co-first author</p>
<section id="summary-2" class="level3" data-number="3.4.1">
<h3 data-number="3.4.1" class="anchored" data-anchor-id="summary-2"><span class="header-section-number">3.4.1</span> Summary</h3>
<p><img src="../papers/mikropml/vignettes/mikropml-logo.png" class="img-fluid" width="120"></p>
<p>Machine learning (ML) for classification and prediction based on a set of features is used to make decisions in healthcare, economics, criminal justice and more. However, implementing an ML pipeline including preprocessing, model selection, and evaluation can be time-consuming, confusing, and difficult. Here, we present <a href="http://www.schlosslab.org/mikropml/"><code>mikropml</code></a> (pronounced “meek-ROPE em el”), an easy-to-use R package that implements ML pipelines using regression, support vector machines, decision trees, random forest, or gradient-boosted trees. The package is available on <a href="https://github.com/SchlossLab/mikropml/">GitHub</a>, <a href="https://cran.r-project.org/package=mikropml">CRAN</a>, and <a href="https://anaconda.org/conda-forge/r-mikropml">conda</a>.</p>
</section>
<section id="statement-of-need-2" class="level3" data-number="3.4.2">
<h3 data-number="3.4.2" class="anchored" data-anchor-id="statement-of-need-2"><span class="header-section-number">3.4.2</span> Statement of need</h3>
<p>Most applications of machine learning (ML) require reproducible steps for data pre-processing, cross-validation, testing, model evaluation, and often interpretation of why the model makes particular predictions. Performing these steps is important, as failure to implement them can result in incorrect and misleading results <span class="citation" data-cites="teschendorff_avoiding_2019 wiens_no_2019">(<a href="#ref-teschendorff_avoiding_2019" role="doc-biblioref">Teschendorff 2019</a>; <a href="#ref-wiens_no_2019" role="doc-biblioref">Wiens et al. 2019</a>)</span>.</p>
<p>Supervised ML is widely used to recognize patterns in large datasets and to make predictions about outcomes of interest. Several packages including <code>caret</code> <span class="citation" data-cites="kuhn_building_2008">(<a href="#ref-kuhn_building_2008" role="doc-biblioref">Kuhn 2008</a>)</span> and <code>tidymodels</code> <span class="citation" data-cites="kuhn_tidymodels_2020">(<a href="#ref-kuhn_tidymodels_2020" role="doc-biblioref">Kuhn, Wickham, and RStudio 2020</a>)</span> in R, <code>scikitlearn</code> <span class="citation" data-cites="pedregosa_scikit-learn_2011">(<a href="#ref-pedregosa_scikit-learn_2011" role="doc-biblioref">Pedregosa et al. 2011</a>)</span> in Python, and the H2O <code>autoML</code> platform <span class="citation" data-cites="h2o_platform">(<a href="#ref-h2o_platform" role="doc-biblioref">H2O.ai 2020</a>)</span> allow scientists to train ML models with a variety of algorithms. While these packages provide the tools necessary for each ML step, they do not implement a complete ML pipeline according to good practices in the literature. This makes it difficult for practitioners new to ML to easily begin to perform ML analyses.</p>
<p>To enable a broader range of researchers to apply ML to their problem domains, we created <a href="https://github.com/SchlossLab/mikropml/"><code>mikropml</code></a>, an easy-to-use R package <span class="citation" data-cites="r_core_team_r_2020">(<a href="#ref-r_core_team_r_2020" role="doc-biblioref">R Core Team 2020</a>)</span> that implements the ML pipeline created by Topçuoğlu <em>et al.</em> <span class="citation" data-cites="topcuoglu_framework_2020">(<a href="#ref-topcuoglu_framework_2020" role="doc-biblioref">Topçuoğlu et al. 2020</a>)</span> in a single function that returns a trained model, model performance metrics and feature importance. <code>mikropml</code> leverages the <code>caret</code> package to support several ML algorithms: linear regression, logistic regression, support vector machines with a radial basis kernel, decision trees, random forest, and gradient boosted trees. It incorporates good practices in ML training, testing, and model evaluation <span class="citation" data-cites="topcuoglu_framework_2020 teschendorff_avoiding_2019">(<a href="#ref-topcuoglu_framework_2020" role="doc-biblioref">Topçuoğlu et al. 2020</a>; <a href="#ref-teschendorff_avoiding_2019" role="doc-biblioref">Teschendorff 2019</a>)</span>. Furthermore, it provides data preprocessing steps based on the FIDDLE (FlexIble Data-Driven pipeLinE) framework outlined in Tang <em>et al.</em> <span class="citation" data-cites="tang_democratizing_2020">(<a href="#ref-tang_democratizing_2020" role="doc-biblioref">Tang et al. 2020</a>)</span> and post-training permutation importance steps to estimate the importance of each feature in the models trained <span class="citation" data-cites="breiman_random_2001 fisher_all_2018">(<a href="#ref-breiman_random_2001" role="doc-biblioref">Breiman 2001</a>; <a href="#ref-fisher_all_2018" role="doc-biblioref">A. Fisher, Rudin, and Dominici 2018</a>)</span>.</p>
<p><code>mikropml</code> can be used as a starting point in the application of ML to datasets from many different fields. It has already been applied to microbiome data to categorize patients with colorectal cancer <span class="citation" data-cites="topcuoglu_framework_2020">(<a href="#ref-topcuoglu_framework_2020" role="doc-biblioref">Topçuoğlu et al. 2020</a>)</span>, to identify differences in genomic and clinical features associated with bacterial infections <span class="citation" data-cites="lapp_machine_2020">(<a href="#ref-lapp_machine_2020" role="doc-biblioref">Lapp et al. 2020</a>)</span>, and to predict gender-based biases in academic publishing <span class="citation" data-cites="hagan_women_2020">(<a href="#ref-hagan_women_2020" role="doc-biblioref">Hagan et al. 2020</a>)</span>.</p>
</section>
<section id="mikropml-package" class="level3" data-number="3.4.3">
<h3 data-number="3.4.3" class="anchored" data-anchor-id="mikropml-package"><span class="header-section-number">3.4.3</span> mikropml package</h3>
<p>The <code>mikropml</code> package includes functionality to preprocess the data, train ML models, evaluate model performance, and quantify feature importance (Figure 1). We also provide <a href="http://www.schlosslab.org/mikropml/articles/index.html">vignettes</a> and an <a href="https://github.com/SchlossLab/mikropml-snakemake-workflow">example Snakemake workflow</a> <span class="citation" data-cites="koster_snakemakescalable_2012">(<a href="#ref-koster_snakemakescalable_2012" role="doc-biblioref">Köster and Rahmann 2012</a>)</span> to showcase how to run an ideal ML pipeline with multiple different train/test data splits. The results can be visualized using helper functions that use <code>ggplot2</code> <span class="citation" data-cites="wickham_ggplot2_2016">(<a href="#ref-wickham_ggplot2_2016" role="doc-biblioref">Wickham 2016</a>)</span>.</p>
<p>While mikropml allows users to get started quickly and facilitates reproducibility, it is not a replacement for understanding the ML workflow which is still necessary when interpreting results <span class="citation" data-cites="pollard_turning_2019">(<a href="#ref-pollard_turning_2019" role="doc-biblioref">Pollard et al. 2019</a>)</span>. To facilitate understanding and enable one to tailor the code to their application, we have heavily commented the code and have provided supporting documentation which can be read <a href="http://www.schlosslab.org/mikropml/">online</a>.</p>
<section id="preprocessing-data" class="level4" data-number="3.4.3.1">
<h4 data-number="3.4.3.1" class="anchored" data-anchor-id="preprocessing-data"><span class="header-section-number">3.4.3.1</span> Preprocessing data</h4>
<p>We provide the function <code>preprocess_data()</code> to preprocess features using several different functions from the <code>caret</code> package. <code>preprocess_data()</code> takes continuous and categorical data, re-factors categorical data into binary features, and provides options to normalize continuous data, remove features with near-zero variance, and keep only one instance of perfectly correlated features. We set the default options based on those implemented in FIDDLE <span class="citation" data-cites="tang_democratizing_2020">(<a href="#ref-tang_democratizing_2020" role="doc-biblioref">Tang et al. 2020</a>)</span>. More details on how to use <code>preprocess_data()</code> can be found in the accompanying <a href="http://www.schlosslab.org/mikropml/articles/preprocess.html">vignette</a>.</p>
</section>
<section id="running-ml" class="level4" data-number="3.4.3.2">
<h4 data-number="3.4.3.2" class="anchored" data-anchor-id="running-ml"><span class="header-section-number">3.4.3.2</span> Running ML</h4>
<p>The main function in mikropml, <code>run_ml()</code>, minimally takes in the model choice and a data frame with an outcome column and feature columns. For model choice, <code>mikropml</code> currently supports logistic and linear regression <span class="citation" data-cites="friedman_regularization_2010">(<code>glmnet</code>: <a href="#ref-friedman_regularization_2010" role="doc-biblioref">Friedman, Hastie, and Tibshirani 2010</a>)</span>, support vector machines with a radial basis kernel <span class="citation" data-cites="karatzoglou_kernlab_2004">(<code>kernlab</code>: <a href="#ref-karatzoglou_kernlab_2004" role="doc-biblioref">Karatzoglou et al. 2004</a>)</span>, decision trees <span class="citation" data-cites="therneau_rpart_2019">(<code>rpart</code>: <a href="#ref-therneau_rpart_2019" role="doc-biblioref">Therneau et al. 2019</a>)</span>, random forest <span class="citation" data-cites="liaw_classication_2002">(<code>randomForest</code>: <a href="#ref-liaw_classication_2002" role="doc-biblioref">Liaw and Wiener 2002</a>)</span>, and gradient-boosted trees <span class="citation" data-cites="chen_xgboost_2020">(<code>xgboost</code>: <a href="#ref-chen_xgboost_2020" role="doc-biblioref">T. Chen et al. 2020</a>)</span>. <code>run_ml()</code> randomly splits the data into train and test sets while maintaining the distribution of the outcomes found in the full dataset. It also provides the option to split the data into train and test sets based on categorical variables (e.g.&nbsp;batch, geographic location, etc.). <code>mikropml</code> uses the <code>caret</code> package <span class="citation" data-cites="kuhn_building_2008">(<a href="#ref-kuhn_building_2008" role="doc-biblioref">Kuhn 2008</a>)</span> to train and evaluate the models, and optionally quantifies feature importance. The output includes the best model built based on tuning hyperparameters in an internal and repeated cross-validation step, model evaluation metrics, and optional feature importances. Feature importances are calculated using a permutation test, which breaks the relationship between the feature and the true outcome in the test data, and measures the change in model performance. This provides an intuitive metric of how individual features influence model performance and is comparable across model types, which is particularly useful for model interpretation <span class="citation" data-cites="topcuoglu_framework_2020">(<a href="#ref-topcuoglu_framework_2020" role="doc-biblioref">Topçuoğlu et al. 2020</a>)</span>. Our <a href="http://www.schlosslab.org/mikropml/articles/introduction.html">introductory vignette</a> contains a comprehensive tutorial on how to use <code>run_ml()</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../papers/mikropml/vignettes/mikRopML-pipeline.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">mikropml pipeline</figcaption>
</figure>
</div>
</section>
<section id="ideal-workflow-for-running-mikropml-with-many-different-traintest-splits" class="level4" data-number="3.4.3.3">
<h4 data-number="3.4.3.3" class="anchored" data-anchor-id="ideal-workflow-for-running-mikropml-with-many-different-traintest-splits"><span class="header-section-number">3.4.3.3</span> Ideal workflow for running mikropml with many different train/test splits</h4>
<p>To investigate the variation in model performance depending on the train and test set used <span class="citation" data-cites="topcuoglu_framework_2020 lapp_machine_2020">(<a href="#ref-topcuoglu_framework_2020" role="doc-biblioref">Topçuoğlu et al. 2020</a>; <a href="#ref-lapp_machine_2020" role="doc-biblioref">Lapp et al. 2020</a>)</span>, we provide examples of how to <code>run_ml()</code> many times with different train/test splits and how to get summary information about model performance on <a href="http://www.schlosslab.org/mikropml/articles/parallel.html">a local computer</a> or on a high-performance computing cluster using a <a href="https://github.com/SchlossLab/mikropml-snakemake-workflow">Snakemake workflow</a>.</p>
</section>
<section id="tuning-visualization" class="level4" data-number="3.4.3.4">
<h4 data-number="3.4.3.4" class="anchored" data-anchor-id="tuning-visualization"><span class="header-section-number">3.4.3.4</span> Tuning &amp; visualization</h4>
<p>One particularly important aspect of ML is hyperparameter tuning. We provide a reasonable range of default hyperparameters for each model type. However practitioners should explore whether that range is appropriate for their data, or if they should customize the hyperparameter range. Therefore, we provide a function <code>plot_hp_performance()</code> to plot the cross-validation performance metric of a single model or models built using different train/test splits. This helps evaluate if the hyperparameter range is being searched exhaustively and allows the user to pick the ideal set. We also provide summary plots of test performance metrics for the many train/test splits with different models using <code>plot_model_performance()</code>. Examples are described in the accompanying <a href="http://www.schlosslab.org/mikropml/articles/tuning.html">vignette on hyperparameter tuning</a>.</p>
</section>
<section id="dependencies" class="level4" data-number="3.4.3.5">
<h4 data-number="3.4.3.5" class="anchored" data-anchor-id="dependencies"><span class="header-section-number">3.4.3.5</span> Dependencies</h4>
<p>mikropml is written in R <span class="citation" data-cites="r_core_team_r_2020">(<a href="#ref-r_core_team_r_2020" role="doc-biblioref">R Core Team 2020</a>)</span> and depends on several packages: <code>dplyr</code> <span class="citation" data-cites="wickham_dplyr_2020">(<a href="#ref-wickham_dplyr_2020" role="doc-biblioref">Wickham et al. 2020</a>)</span>, <code>rlang</code> <span class="citation" data-cites="henry_rlang_2020">(<a href="#ref-henry_rlang_2020" role="doc-biblioref">Henry, Wickham, and RStudio 2020</a>)</span> and <code>caret</code> <span class="citation" data-cites="kuhn_building_2008">(<a href="#ref-kuhn_building_2008" role="doc-biblioref">Kuhn 2008</a>)</span>. The ML algorithms supported by <code>mikropml</code> require: <code>glmnet</code> <span class="citation" data-cites="friedman_regularization_2010">(<a href="#ref-friedman_regularization_2010" role="doc-biblioref">Friedman, Hastie, and Tibshirani 2010</a>)</span>, <code>e1071</code> <span class="citation" data-cites="meyer_e1071_2020">(<a href="#ref-meyer_e1071_2020" role="doc-biblioref">Meyer et al. 2020</a>)</span>, and <code>MLmetrics</code> <span class="citation" data-cites="yan_mlmetrics_2016">(<a href="#ref-yan_mlmetrics_2016" role="doc-biblioref">Yan 2016</a>)</span> for logistic regression, <code>rpart2</code> <span class="citation" data-cites="therneau_rpart_2019">(<a href="#ref-therneau_rpart_2019" role="doc-biblioref">Therneau et al. 2019</a>)</span> for decision trees, <code>randomForest</code> <span class="citation" data-cites="liaw_classication_2002">(<a href="#ref-liaw_classication_2002" role="doc-biblioref">Liaw and Wiener 2002</a>)</span> for random forest, <code>xgboost</code> <span class="citation" data-cites="chen_xgboost_2020">(<a href="#ref-chen_xgboost_2020" role="doc-biblioref">T. Chen et al. 2020</a>)</span> for xgboost, and <code>kernlab</code> <span class="citation" data-cites="karatzoglou_kernlab_2004">(<a href="#ref-karatzoglou_kernlab_2004" role="doc-biblioref">Karatzoglou et al. 2004</a>)</span> for support vector machines. We also allow for parallelization of cross-validation and other steps using the <code>foreach</code>, <code>doFuture</code>, <code>future.apply</code>, and <code>future</code> packages <span class="citation" data-cites="bengtsson_futureapply_2020">(<a href="#ref-bengtsson_futureapply_2020" role="doc-biblioref">Bengtsson and Team 2020</a>)</span>. Finally, we use <code>ggplot2</code> for plotting <span class="citation" data-cites="wickham_ggplot2_2016">(<a href="#ref-wickham_ggplot2_2016" role="doc-biblioref">Wickham 2016</a>)</span>.</p>
</section>
</section>
<section id="acknowledgments" class="level3" data-number="3.4.4">
<h3 data-number="3.4.4" class="anchored" data-anchor-id="acknowledgments"><span class="header-section-number">3.4.4</span> Acknowledgments</h3>
<p>We thank members of the Schloss Lab who participated in code clubs related to the initial development of the pipeline, made documentation improvements, and provided general feedback. We also thank Nick Lesniak for designing the mikropml logo.</p>
<p>We thank the US Research Software Sustainability Institute (NSF #1743188) for providing training to KLS at the Winter School in Research Software Engineering.</p>
</section>
<section id="funding-2" class="level3" data-number="3.4.5">
<h3 data-number="3.4.5" class="anchored" data-anchor-id="funding-2"><span class="header-section-number">3.4.5</span> Funding</h3>
<p>Salary support for PDS came from NIH grant 1R01CA215574. KLS received support from the NIH Training Program in Bioinformatics (T32 GM070449). ZL received support from the National Science Foundation Graduate Research Fellowship Program under Grant No.&nbsp;DGE 1256260. Any opinions, findings, and conclusions or recommendations expressed in this material are those of the authors and do not necessarily reflect the views of the National Science Foundation.</p>
</section>
<section id="author-contributions-2" class="level3" data-number="3.4.6">
<h3 data-number="3.4.6" class="anchored" data-anchor-id="author-contributions-2"><span class="header-section-number">3.4.6</span> Author contributions</h3>
<p>BDT, ZL, and KLS contributed equally. Author order among the co-first authors was determined by time since joining the project.</p>
<p>BDT, ZL, and KLS conceptualized the study and wrote the code. KLS structured the code in R package form. BDT, ZL, JW, and PDS developed methodology. PDS, ES, and JW supervised the project. BDT, ZL, and KLS wrote the original draft. All authors reviewed and edited the manuscript.</p>
</section>
<section id="conflicts-of-interest-2" class="level3" data-number="3.4.7">
<h3 data-number="3.4.7" class="anchored" data-anchor-id="conflicts-of-interest-2"><span class="header-section-number">3.4.7</span> Conflicts of interest</h3>
<p>None.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-becker_responding_2016" class="csl-entry" role="listitem">
Becker, Erin. 2016. <span>“Responding to Your <span>Learners</span>.”</span> <em>Data Carpentry</em>. <a href="https://datacarpentry.org/blog/2016/09/formative-assessment" class="uri">https://datacarpentry.org/blog/2016/09/formative-assessment</a>.
</div>
<div id="ref-benbow_gender_2016" class="csl-entry" role="listitem">
Benbow, Ross J., and Erika Vivyan. 2016. <span>“Gender and <span>Belonging</span> in <span>Undergraduate Computer Science</span>: <span>A Comparative Case Study</span> of <span>Student Experiences</span> in <span>Gateway Courses</span>.”</span>
</div>
<div id="ref-bengtsson_futureapply_2020" class="csl-entry" role="listitem">
Bengtsson, Henrik, and R Core Team. 2020. <span>“Future.apply: <span>Apply Function</span> to <span>Elements</span> in <span>Parallel</span> Using <span>Futures</span>.”</span>
</div>
<div id="ref-bonham_women_2017" class="csl-entry" role="listitem">
Bonham, Kevin S., and Melanie I. Stefan. 2017. <span>“Women Are Underrepresented in Computational Biology: <span>An</span> Analysis of the Scholarly Literature in Biology, Computer Science and Computational Biology.”</span> <em>PLOS Computational Biology</em> 13 (10): e1005134. <a href="https://doi.org/10.1371/journal.pcbi.1005134">https://doi.org/10.1371/journal.pcbi.1005134</a>.
</div>
<div id="ref-breiman_random_2001" class="csl-entry" role="listitem">
Breiman, Leo. 2001. <span>“Random Forests.”</span> <em>Machine Learning</em> 45 (1): 5–32. <a href="https://doi.org/10.1023/A:1010933404324">https://doi.org/10.1023/A:1010933404324</a>.
</div>
<div id="ref-chen_online_2020" class="csl-entry" role="listitem">
Chen, Daniel. 2020. <span>“Online <span>Workshop Logistics</span> and <span>Screen Layouts</span>.”</span> <em>The Carpentries</em>. <a href="https://carpentries.org/blog/2020/06/online-workshop-logistics-and_screen-layouts/" class="uri">https://carpentries.org/blog/2020/06/online-workshop-logistics-and_screen-layouts/</a>.
</div>
<div id="ref-chen_xgboost_2020" class="csl-entry" role="listitem">
Chen, Tianqi, Tong He, Michael Benesty, Vadim Khotilovich, Yuan Tang, Hyunsu Cho, Kailong Chen, et al. 2020. <span>“Xgboost: <span>Extreme Gradient Boosting</span>.”</span>
</div>
<div id="ref-fisher_all_2018" class="csl-entry" role="listitem">
Fisher, Aaron, Cynthia Rudin, and Francesca Dominici. 2018. <span>“All Models Are Wrong, but Many Are Useful: <span>Learning</span> a Variable’s Importance by Studying an Entire Class of Prediction Models Simultaneously.”</span>
</div>
<div id="ref-fisher_better_2013" class="csl-entry" role="listitem">
Fisher, Douglas, and Nancy Frey. 2013. <em>Better <span>Learning Through Structured Teaching</span>: <span>A Framework</span> for the <span>Gradual Release</span> of <span>Responsibility</span>, 2nd <span>Edition</span></em>.
</div>
<div id="ref-friedman_regularization_2010" class="csl-entry" role="listitem">
Friedman, Jerome H., Trevor Hastie, and Rob Tibshirani. 2010. <span>“Regularization <span>Paths</span> for <span>Generalized Linear Models</span> via <span>Coordinate Descent</span>.”</span> <em>Journal of Statistical Software</em> 33 (1): 1–22. <a href="https://doi.org/10.18637/jss.v033.i01">https://doi.org/10.18637/jss.v033.i01</a>.
</div>
<div id="ref-hq_gallery_2021" class="csl-entry" role="listitem">
Girls Who Code HQ. 2021. <span>“Girls <span>Who Code Project Gallery</span>.”</span> https://hq.girlswhocode.com/project-gallery.
</div>
<div id="ref-h2o_platform" class="csl-entry" role="listitem">
H2O.ai. 2020. <em><span>H2O</span>: <span>Scalable</span> Machine Learning Platform</em>. Manual.
</div>
<div id="ref-hagan_women_2020" class="csl-entry" role="listitem">
Hagan, Ada K., Begüm D. Topçuoğlu, Mia E. Gregory, Hazel A. Barton, and Patrick D. Schloss. 2020. <span>“Women <span>Are Underrepresented</span> and <span>Receive Differential Outcomes</span> at <span>ASM Journals</span>: A <span>Six</span>-<span>Year Retrospective Analysis</span>.”</span> <em>mBio</em> 11 (6). <a href="https://doi.org/10.1128/mBio.01680-20">https://doi.org/10.1128/mBio.01680-20</a>.
</div>
<div id="ref-hannay_effectiveness_2009" class="csl-entry" role="listitem">
Hannay, Jo E., Tore Dybå, Erik Arisholm, and Dag I. K. Sjøberg. 2009. <span>“The Effectiveness of Pair Programming: <span>A</span> Meta-Analysis.”</span> <em>Information and Software Technology</em>, Special <span>Section</span>: <span>Software Engineering</span> for <span>Secure Systems</span>, 51 (7): 1110–22. <a href="https://doi.org/10.1016/j.infsof.2009.02.001">https://doi.org/10.1016/j.infsof.2009.02.001</a>.
</div>
<div id="ref-harris_array_2020" class="csl-entry" role="listitem">
Harris, Charles R., K. Jarrod Millman, Stéfan J. van der Walt, Ralf Gommers, Pauli Virtanen, David Cournapeau, Eric Wieser, et al. 2020. <span>“Array Programming with <span>NumPy</span>.”</span> <em>Nature</em> 585 (7825): 357–62. <a href="https://doi.org/10.1038/s41586-020-2649-2">https://doi.org/10.1038/s41586-020-2649-2</a>.
</div>
<div id="ref-henry_rlang_2020" class="csl-entry" role="listitem">
Henry, Lionel, Hadley Wickham, and RStudio. 2020. <span>“Rlang: <span>Functions</span> for <span>Base Types</span> and <span>Core R</span> and ’<span>Tidyverse</span>’ <span>Features</span>.”</span>
</div>
<div id="ref-hill_why_2010" class="csl-entry" role="listitem">
Hill, Catherine, Christianne Corbett, and Andresse St. Rose. 2010. <em>Why so Few? Women in Science, Technology, Engineering, and Mathematics</em>. <span>Washington, D.C</span>: <span>AAUW</span>.
</div>
<div id="ref-hunter_matplotlib_2007" class="csl-entry" role="listitem">
Hunter, John D. 2007. <span>“Matplotlib: <span>A 2D</span> Graphics Environment.”</span> <em>Computing in Science Engineering</em> 9 (3): 90–95. <a href="https://doi.org/10.1109/MCSE.2007.55">https://doi.org/10.1109/MCSE.2007.55</a>.
</div>
<div id="ref-karatzoglou_kernlab_2004" class="csl-entry" role="listitem">
Karatzoglou, Alexandros, Alexandros Smola, Kurt Hornik, and Achim Zeileis. 2004. <span>“Kernlab - <span>An S4 Package</span> for <span>Kernel Methods</span> in <span>R</span>.”</span> <em>Journal of Statistical Software</em> 11 (1): 1–20. <a href="https://doi.org/10.18637/jss.v011.i09">https://doi.org/10.18637/jss.v011.i09</a>.
</div>
<div id="ref-koster_snakemakescalable_2012" class="csl-entry" role="listitem">
Köster, Johannes, and Sven Rahmann. 2012. <span>“Snakemakea Scalable Bioinformatics Workflow Engine.”</span> <em>Bioinformatics</em> 28 (19): 2520–22. <a href="https://doi.org/10.1093/bioinformatics/bts480">https://doi.org/10.1093/bioinformatics/bts480</a>.
</div>
<div id="ref-kuhn_building_2008" class="csl-entry" role="listitem">
Kuhn, Max. 2008. <span>“Building <span>Predictive Models</span> in <span>R Using</span> the Caret <span>Package</span>.”</span> <em>Journal of Statistical Software</em> 28 (1): 1–26. <a href="https://doi.org/10.18637/jss.v028.i05">https://doi.org/10.18637/jss.v028.i05</a>.
</div>
<div id="ref-kuhn_tidymodels_2020" class="csl-entry" role="listitem">
Kuhn, Max, Hadley Wickham, and RStudio. 2020. <span>“Tidymodels: <span>Easily Install</span> and <span>Load</span> the ’<span>Tidymodels</span>’ <span>Packages</span>.”</span>
</div>
<div id="ref-lapp_machine_2020" class="csl-entry" role="listitem">
Lapp, Zena, Jennifer Han, Jenna Wiens, Ellie JC Goldstein, Ebbing Lautenbach, and Evan Snitkin. 2020. <span>“Machine Learning Models to Identify Patient and Microbial Genetic Factors Associated with Carbapenem-Resistant <span>Klebsiella</span> Pneumoniae Infection.”</span> <em>medRxiv</em>, 2020.07.06.20147306. <a href="https://doi.org/10.1101/2020.07.06.20147306">https://doi.org/10.1101/2020.07.06.20147306</a>.
</div>
<div id="ref-lapp_developing_2021" class="csl-entry" role="listitem">
Lapp, Zena, Kelly L Sovacool, Nicholas A Lesniak, Dana King, Catherine Barnier, Matthew Flickinger, Jule Krüger, et al. 2021. <span>“Developing and Deploying an Integrated Workshop Curriculum Teaching Computational Skills for Reproducible Research.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.06.15.448091">https://doi.org/10.1101/2021.06.15.448091</a>.
</div>
<div id="ref-liaw_classication_2002" class="csl-entry" role="listitem">
Liaw, Andy, and Matthew Wiener. 2002. <span>“Classification and <span>Regression</span> by <span class="nocase">randomForest</span>”</span> 2: 5.
</div>
<div id="ref-meyer_e1071_2020" class="csl-entry" role="listitem">
Meyer, David, Evgenia Dimitriadou, Kurt Hornik, Andreas Weingessel, Friedrich Leisch, Chih-Chung Chang (libsvm C++-code), and Chih-Chen Lin (libsvm C++-code). 2020. <span>“E1071: <span>Misc Functions</span> of the <span>Department</span> of <span>Statistics</span>, <span>Probability Theory Group</span> (<span>Formerly</span>: <span>E1071</span>), <span>TU Wien</span>.”</span>
</div>
<div id="ref-nces_digest_2012" class="csl-entry" role="listitem">
National Center for Education Statistics. 2012. <span>“Digest of <span>Education Statistics</span>.”</span> https://nces.ed.gov/programs/digest/d12/tables/dt12_349.asp; <span>National Center for Education Statistics</span>.
</div>
<div id="ref-nederbragt_ten_2020" class="csl-entry" role="listitem">
Nederbragt, Alexander, Rayna Michelle Harris, Alison Presmanes Hill, and Greg Wilson. 2020. <span>“Ten Quick Tips for Teaching with Participatory Live Coding.”</span> <em>PLOS Computational Biology</em> 16 (9): e1008090. <a href="https://doi.org/10.1371/journal.pcbi.1008090">https://doi.org/10.1371/journal.pcbi.1008090</a>.
</div>
<div id="ref-pedregosa_scikit-learn_2011" class="csl-entry" role="listitem">
Pedregosa, Fabian, Gaël Varoquaux, Alexandre Gramfort, Vincent Michel, Bertrand Thirion, Olivier Grisel, Mathieu Blondel, et al. 2011. <span>“Scikit-Learn: <span>Machine Learning</span> in <span>Python</span>.”</span> <em>Journal of Machine Learning Research</em> 12 (85): 2825–30.
</div>
<div id="ref-pollard_turning_2019" class="csl-entry" role="listitem">
Pollard, Tom J., Irene Chen, Jenna Wiens, Steven Horng, Danny Wong, Marzyeh Ghassemi, Heather Mattie, Emily Lindemer, and Trishan Panch. 2019. <span>“Turning the Crank for Machine Learning: Ease, at What Expense?”</span> <em>The Lancet Digital Health</em> 1 (5): e198–99. <a href="https://doi.org/10.1016/S2589-7500(19)30112-8">https://doi.org/10.1016/S2589-7500(19)30112-8</a>.
</div>
<div id="ref-noauthor_pypl_nodate" class="csl-entry" role="listitem">
<span>“<span>PYPL PopularitY</span> of <span>Programming Language</span> Index.”</span> n.d. https://pypl.github.io/PYPL.html.
</div>
<div id="ref-r_core_team_r_2020" class="csl-entry" role="listitem">
R Core Team. 2020. <em>R: <span>A</span> Language and Environment for Statistical Computing</em>. Manual. <span>Vienna, Austria</span>: <span>R Foundation for Statistical Computing</span>.
</div>
<div id="ref-robinson_teach_2017" class="csl-entry" role="listitem">
Robinson, David. 2017. <span>“Teach the Tidyverse to Beginners.”</span> <em>Variance Explained</em>. <a href="http://varianceexplained.org/r/teach-tidyverse/" class="uri">http://varianceexplained.org/r/teach-tidyverse/</a>.
</div>
<div id="ref-saujani_girls_2015" class="csl-entry" role="listitem">
Saujani, Reshma. 2015. <span>“Girls <span>Who Code</span>: <span>Annual Report</span> 2015.”</span> http://girlswhocode.com/2015report/.
</div>
<div id="ref-stansell_breaking_2019" class="csl-entry" role="listitem">
Stansell, Amanda. 2019. <span>“Breaking <span>Down</span> the 50 <span>Best Jobs</span> in <span>America</span> for 2019 - <span>Glassdoor</span>.”</span> <em>Glassdoor Economic Research</em>. https://www.glassdoor.com/research/best-jobs-2019/.
</div>
<div id="ref-tang_democratizing_2020" class="csl-entry" role="listitem">
Tang, Shengpu, Parmida Davarmanesh, Yanmeng Song, Danai Koutra, Michael W. Sjoding, and Jenna Wiens. 2020. <span>“Democratizing <span>EHR</span> Analyses with <span>FIDDLE</span>: A Flexible Data-Driven Preprocessing Pipeline for Structured Clinical Data.”</span> <em>J Am Med Inform Assoc</em>. <a href="https://doi.org/10.1093/jamia/ocaa139">https://doi.org/10.1093/jamia/ocaa139</a>.
</div>
<div id="ref-teschendorff_avoiding_2019" class="csl-entry" role="listitem">
Teschendorff, Andrew E. 2019. <span>“Avoiding Common Pitfalls in Machine Learning Omic Data Science.”</span> <em>Nature Materials</em> 18 (5): 422–27. <a href="https://doi.org/10.1038/s41563-018-0241-z">https://doi.org/10.1038/s41563-018-0241-z</a>.
</div>
<div id="ref-the_carpentries_live_2018" class="csl-entry" role="listitem">
The Carpentries. 2018a. <span>“Live <span>Coding</span> Is a <span>Skill</span>.”</span> <em>Instructor Training</em>. <a href="https://carpentries.github.io/instructor-training/14-live/#sticky-notes" class="uri">https://carpentries.github.io/instructor-training/14-live/#sticky-notes</a>.
</div>
<div id="ref-the_carpentries_carpentries_2018" class="csl-entry" role="listitem">
———. 2018b. <span>“The <span>Carpentries Handbook</span>.”</span> <a href="https://docs.carpentries.org/index.html" class="uri">https://docs.carpentries.org/index.html</a>.
</div>
<div id="ref-therneau_rpart_2019" class="csl-entry" role="listitem">
Therneau, Terry, Beth Atkinson, Brian Ripley (producer of the initial R. port, and maintainer 1999-2017). 2019. <span>“Rpart: <span>Recursive Partitioning</span> and <span>Regression Trees</span>.”</span>
</div>
<div id="ref-topcuoglu_framework_2020" class="csl-entry" role="listitem">
Topçuoğlu, Begüm D., Nicholas A. Lesniak, Mack T. Ruffin, Jenna Wiens, and Patrick D. Schloss. 2020. <span>“A <span>Framework</span> for <span>Effective Application</span> of <span>Machine Learning</span> to <span>Microbiome</span>-<span>Based Classification Problems</span>.”</span> <em>mBio</em> 11 (3). <a href="https://doi.org/10.1128/mBio.00434-20">https://doi.org/10.1128/mBio.00434-20</a>.
</div>
<div id="ref-waskom_seaborn_2021" class="csl-entry" role="listitem">
Waskom, Michael. 2021. <span>“Seaborn: Statistical Data Visualization.”</span> <em>JOSS</em> 6 (60): 3021. <a href="https://doi.org/10.21105/joss.03021">https://doi.org/10.21105/joss.03021</a>.
</div>
<div id="ref-wickham_ggplot2_2016" class="csl-entry" role="listitem">
Wickham, Hadley. 2016. <em>Ggplot2: <span>Elegant Graphics</span> for <span>Data Analysis</span></em>. Use <span>R</span>! <span>Cham</span>: <span>Springer International Publishing</span>. <a href="https://doi.org/10.1007/978-3-319-24277-4">https://doi.org/10.1007/978-3-319-24277-4</a>.
</div>
<div id="ref-wickham_welcome_2019" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the <span>Tidyverse</span>.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-wickham_dplyr_2020" class="csl-entry" role="listitem">
Wickham, Hadley, Romain François, Lionel Henry, Kirill Müller, and RStudio. 2020. <span>“Dplyr: <span>A Grammar</span> of <span>Data Manipulation</span>.”</span>
</div>
<div id="ref-wiens_no_2019" class="csl-entry" role="listitem">
Wiens, Jenna, Suchi Saria, Mark Sendak, Marzyeh Ghassemi, Vincent X. Liu, Finale Doshi-Velez, Kenneth Jung, et al. 2019. <span>“Do No Harm: A Roadmap for Responsible Machine Learning for Health Care.”</span> <em>Nat. Med.</em> 25 (9): 1337–40. <a href="https://doi.org/10.1038/s41591-019-0548-6">https://doi.org/10.1038/s41591-019-0548-6</a>.
</div>
<div id="ref-wilson_software_2016" class="csl-entry" role="listitem">
Wilson, Greg. 2016. <span>“Software <span>Carpentry</span>: Lessons Learned.”</span> <em>F1000Res</em> 3: 62. <a href="https://doi.org/10.12688/f1000research.3-62.v2">https://doi.org/10.12688/f1000research.3-62.v2</a>.
</div>
<div id="ref-xie_r_2018" class="csl-entry" role="listitem">
Xie, Yihui, J. J. Allaire, and Garrett Grolemund. 2018. <em>R <span>Markdown</span>: <span>The Definitive Guide</span></em>. <span>Taylor &amp; Francis, CRC Press</span>.
</div>
<div id="ref-yan_mlmetrics_2016" class="csl-entry" role="listitem">
Yan, Yachen. 2016. <span>“<span>MLmetrics</span>: <span>Machine Learning Evaluation Metrics</span>.”</span>
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/optifit.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">OptiFit: an improved method for fitting amplicon sequences to existing OTUs</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/cdiff.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Predicting <em>C. difficile</em> infection severity from the taxonomic composition of the gut microbiome</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>