```{r mikropml_downloads}
schtools::set_knitr_opts()
library(cranlogs)
library(dplyr)
cran_downloads <- cran_downloads(package = "mikropml", 
                                 from = "2020-11-23", to = "2023-06-08") %>% 
  pull(count) %>% sum()
conda_downloads <- 24727
```

# Discussion

## Major contributions

This dissertation introduces two new tools that improve ML capabilities for
microbiome research and beyond, applies ML with microbiome data to CDI severity
prediction, and introduces two new educational resources that teach coding for
data science to young audiences and scientists.
For all of the analyses described in this dissertation, the complete software
workflows and dependencies required to reproduce the results are publicly
available with open source licenses so that anyone can reproduce, replicate, or
build upon our work.
The impact of this work spans microbial ecology, gut microbiome research,
applied machine learning, and data science education.

### Novel method for reference-based OTU clustering

OptiFit is a novel OTU clustering method that enables high quality OTUs for ML
workflows and other applications where consistent OTUs are required.
Prior to the development of OptiFit, the only option for researchers who wanted
to deploy OTU-based ML models was to cluster both the training set and external
validation sets to the same database using a closed-reference clustering method.
Existing tools for reference-based clustering against databases produce lower
quality OTUs than _de novo_ clustering with OptiClust.
However, _de novo_ clustering results in slightly different OTU assignments when
adding new sequences, thus models trained on one dataset could not be deployed
on new data due to incompatible features.
Now with OptiFit, an initial dataset can be clustered _de novo_ with OptiClust
and then used to train a model, then new sequences from an external validation
set can be fit to the OTUs from the training data prior to deploying the model
on the new dataset.
A follow-up paper demonstrated the suitability of OptiFit for this very task
on a colorectal cancer dataset to distinguish patients with screen-relevant 
neoplasias from normal controls [@armour_machine_2023].
OptiFit opens a new door for microbial ecologists to deploy ML models using
higher quality OTUs than were possible before.

### Microbiome models for prediction of severe CDI outcomes

Prior studies to date have trained models to predict severe CDI outcomes using
routine clinical data, selected serum biomarkers, curated variables from EHR
data, or entire EHRs.
However, none have focused on using the initial taxonomic composition of the gut
microbiome to predict CDI severity, despite ample evidence for a link between
dysbiosis and _C. difficile_ colonization, infection, and recurrence.
We trained models on OTU relative abundances collected on the day of CDI
diagnosis to predict four different definitions of severity.
Models trained to predict the pragmatic severity definition performed best, as
this definition uses as much data as possible while also using physicians'
determinations of whether severe outcomes were CDI-attributable when available.
While these models did not outperform prior EHR-based models extracted two days
after diagnosis, the pragmatic severity models matched the performance of
EHR-based models from the day of diagnosis.
These results provide an initial exploration of the utility of OTU-based models
for predicting CDI severity, and they may become more clinically relevant in the
future as new evidence emerges of efficacious treatments for preventing severity.

### Educational resources

In chapter 4.2, we introduced a new curriculum to teach introductory Python for
data science via live-coding or a flipped classroom format.
We deployed the curriculum for in-person and virtual Girls Who Code clubs during
a three year period with high school students as the audience.
The curriculum takes students from having no knowledge of programming to being
able to analyze a real-world dataset and present their findings to the group.
In a post-survey, students overwhelmingly reported that they improved their
Python programming skills, problem solving and critical thinking, collaboration
with others, and self-confidence.
Not only were the students we taught positively impacted; our curriculum is
free and available with an open source license so any other educators can use
our curriculum or modify it for their own needs.
This curriculum is continually improved upon and is still in use for the chapter
of Girls Who Code at the University of Michigan Department of Computational
Medicine and Bioinformatics.

In chapter 4.3, we introduced a new curriculum to teach coding for reproducible
research practices to scientists and other researchers in an academic setting.
The Carpentries materials that inspired us taught three topics in a disparate
manner: introductory R programming, the Unix shell, and version control with
git and GitHub.
Our curriculum covers these topics in an integrated manner so that learners
understand how they are used together in practice.
We piloted the curriculum in a virtual workshop and assessed our work with a
post-workshop survey.
On average, learners reported that they felt more confident writing programs,
using programming to work with data, overcoming problems while programming, and
searching for answers to technical questions online.
This curriculum is still in use today for Carpentries workshops at the
University of Michigan and is freely available with an open source license for
anyone to use and modify.

### Software

In chapter 4.4, we introduced a tool that integrates current best practices
for ML in a user-friendly R package.
Our goal was to enable researchers who are novices in ML to train and evaluate
models with guard rails to prevent common pitfalls, while allowing experienced
users to tailor the package for advanced needs.
At the time of this writing, mikropml has been downloaded `r cran_downloads`
times from the Comprehensive R Archive Network and `r conda_downloads` times
from the Anaconda package manager, suggesting a healthy user base.
The reach of mikropml has expanded outside of our immediate scientific network
and into fields spanning gut microbiome research, microbial ecology, public
health, and environmental research.
Rather than write code intended for one-time-use-only to conduct the ML analyses
shown in this work, we chose to bundle our methods into a package for others to
reuse for their own research.
As a result, our efforts have contributed directly to the greater scientific
endeavor, with 18 citations to date of the mikropml publication in the Journal
of Open Source Software.
<!-- mikropml snakemake workflow, schtools, mothur snakemake workflow -->

## Future work

### Integrate microbiota with clinical factors for improved CDI severity prediction

Our OTU-based models described in chapter 3 were trained on a different dataset
as the EHR-based models we compared them to.
Since the different datasets have different proportions of severe cases,
precision and AUPRC are not directly comparable.
While AUROC has the same baseline regardless of the dataset and is thus always
directly comparable, it is not as useful for rare outcomes because the model
may identify many true negatives but few true positives and yet report a high
AUROC.
A more salient comparison would train models on the same cohort of patients
using either OTUs, EHRs, or both in order to determine which approach leads to
the best performance in terms of AUPRC.
However, to demonstrate clinical value, it is not enough to simply show that one
modeling approach outperforms another.
How a model might improve clinical practice if it were deployed must be
considered.
This especially relates to the treatment options available along with their
potential risks, which influences which performance metrics are most meaningful.
A large increase in AUROC, AUPRC, or other metrics may or may not translate to a
large increase in benefit to to patients.
In situations where predicting a severe case may lead clinicians to choose a
treatment option that has an established record of safety, such as oral
fidaxomicin instead of vancomycin, false positives are tolerable and a lower
precision is acceptable.
On the other hand, if new evidence were to emerge of a treatment preventing
severe outcomes but with substantial risk of negative side effects, false
positives would be less acceptable and a higher precision would be required.

For the analysis of potential clinical value, we reported the precision at the
95th percentile of risk, which is the decision threshold where 5% of cases are
predicted to be positive and would thus undergo a different treatment in order
to prevent the adverse outcome from occurring.
Choosing this threshold allowed for comparison to the previously published EHR
models which reported precision at that threshold.
However, rather than evaluating performance at a single threshold,
we could extend this across a range of thresholds.
Decision threshold analysis would explore how the confusion matrix varies
across a range of thresholds for models of interest.
We could then compare the net benefit, NNS, or other metrics for different
modeling approaches, as their relative performance may vary across decision
thresholds.
A model based on only OTUs may perform optimally at a different threshold than
a model based on only EHRs, and different thresholds could be selected for model
deployment depending on the importance of recall versus precision for the
alternative treatment being considered by clinicians.

The cost of treatment is a significant factor that influences the practicality
of model deployment.

bezlotoxumab shown to prevent systemic organ damage in a mouse study [@mileto_bezlotoxumab_2022].


### Beyond taxonomic composition

consider functional potential and actual function

Efforts to find consistent changes in taxonomic composition of microbiomes
between normal and dysbiotic states have found mixed success, in part because
interpersonal variability in taxonomic composition sometimes exceeds the
variability between disease states.
Variability of microbiome composition between individuals with the same disease
status may be explained by functional redundancy, where different microbial
species carry out the same functions and thus can replace each other with
little effect on the overall function of the community.

Sequencing whole metagenomes to identify the genes present and annotate known
gene functions is commonly used to build a profile of functional potential of
the microbiome.
Combining taxonomic composition from OTUs with functional potential from
metagenomes allows one to characterize functional redundancy across communities,
where communities with similar functional potential have different taxonomic
composition.
Untargeted mass spectrometry or transcriptomics can validate the functional
potential characterized from metagenomics by identifying metabolites or gene
products that are active in a community,
thus painting a more precise picture of active microbial functions.
... investigate the impacts of taking functional redundancy and active metabolites into account for microbiome analyses.
incorporate functional composition from metagenomic data to improve the
performance of machine learning models...

Previous studies have built OTU-based machine learning models to classify stool
samples as normal or cancerous to serve as a less invasive diagnostic tool for
CRC than colonoscopy, but have only achieved modest performance.
Incorporating the known functional potential of the microbiome from metagenomic
data may help account for functional redundancy and improve the performance of
OTU-based models in classifying CRC, predicting CDI severity, or other 
microbiome modeling problems.

### Democratizing the democratizers

did i seriously just write that

## Conclusions

TODO


The ultimate goal of CDI severity prediction models is to help clinicians
identify early on which patients are at risk of experiencing a severe outcome
so they can tailor treatments to prevent the outcome from ever occurring.
