

Assigning amplicon sequences to operational taxonomic units (OTUs) is an
important step in characterizing microbial communities across large
datasets. A notable difference between *de novo* clustering and
database-dependent reference clustering methods is that OTU assignments
from *de novo* methods may change when new sequences are added. However,
one may wish to incorporate new samples to previously clustered datasets
without clustering all sequences again, such as when comparing across
datasets or deploying machine learning models. Existing reference-based
methods produce consistent OTUs, but only consider the similarity of
each query sequence to a single reference sequence in an OTU, resulting
in assignments that are worse than those generated by *de novo* methods.
To provide an efficient method to fit sequences to existing OTUs, we
developed the OptiFit algorithm. Inspired by the *de novo* OptiClust
algorithm, OptiFit considers the similarity of all pairs of reference
and query sequences to produce OTUs of the best possible quality. We
tested OptiFit using four datasets with two strategies: 1) clustering to
a reference database or 2) splitting the dataset into a reference and
query set, clustering the references using OptiClust, then clustering
the queries to the references. The result is an improved implementation
of reference-based clustering. OptiFit produces similar quality OTUs as
OptiClust at faster speeds when using the split dataset strategy.
OptiFit provides a suitable option for users requiring consistent OTU
assignments at the same quality afforded by *de novo* clustering
methods.

*Clostridioides difficile* infection (CDI) can lead to adverse outcomes
including ICU admission, colectomy, and death. The composition of the
gut microbiome plays an important role in determining colonization
resistance and clearance upon exposure to *C. difficile*. We
investigated whether machine learning (ML) models trained on 16S rRNA
gene amplicon sequences from gut microbiota extracted from 1,277 patient
stool samples on the day of CDI diagnosis could predict which CDI cases
led to severe outcomes. We then trained ML models to predict CDI
severity on OTU relative abundances according to four different severity
definitions: the IDSA severity score on the day of diagnosis, all-cause
adverse outcomes within 30 days, adverse outcomes confirmed as
attributable to CDI via chart review, and a pragmatic definition that
uses the attributable definition when available and otherwise uses the
all-cause definition. The models predicting pragmatic severity performed
best, suggesting that while chart review is valuable to verify the cause
of complications, including as many samples as possible is indispensable
for training performant models on imbalanced datasets. Permutation
importance identified *Enterococcus* as the most important OTU for model
performance, and increased relative abundance of *Enterococcus* was
associated with severe outcomes. Finally, we evaluated the potential
clinical value of the OTU-based models and found similar performance
compared to prior models based on Electronic Health Records. The modest
performance of the OTU-based models represents a step toward the goal of
deploying models to inform clinical decisions and ultimately improve CDI
outcomes.
